// zigbee-herdsman-converters/dist/devices/alecto.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.definitions = void 0;
const exposes = __importStar(require("../lib/exposes"));
const legacy = __importStar(require("../lib/legacy"));
const tuya = __importStar(require("../lib/tuya"));
const e = exposes.presets;
const ea = exposes.access;
exports.definitions = [
    {
        fingerprint: [...tuya.fingerprint("daqwrsj\u0000", ["_TYST11_8daqwrsj"]), ...tuya.fingerprint("TS0601", ["_TZE200_8daqwrsj"])],
        model: "SMART-HEAT10",
        vendor: "Alecto",
        description: "Radiator valve with thermostat",
        fromZigbee: [legacy.fz.tuya_thermostat],
        meta: {
            tuyaThermostatSystemMode: legacy.thermostatSystemModes4,
            tuyaThermostatPreset: legacy.thermostatPresets,
            tuyaThermostatPresetToSystemMode: legacy.thermostatSystemModes4,
        },
        toZigbee: [
            legacy.tz.tuya_thermostat_child_lock,
            legacy.tz.siterwell_thermostat_window_detection,
            legacy.tz.tuya_thermostat_current_heating_setpoint,
            legacy.tz.tuya_thermostat_system_mode,
        ],
        exposes: [
            e.child_lock(),
            e.window_detection(),
            e.battery(),
            e
                .climate()
                .withSetpoint("current_heating_setpoint", 5, 30, 0.5, ea.STATE_SET)
                .withLocalTemperature(ea.STATE)
                .withSystemMode(["off", "auto", "heat"], ea.STATE_SET),
        ],
    },
    {
        fingerprint: [...tuya.fingerprint("tbrwrfv\u0000", ["_TYST11_qtbrwrfv"]), ...tuya.fingerprint("TS0601", ["_TZE200_qtbrwrfv"])],
        model: "SMART-SMOKE10",
        vendor: "Alecto",
        description: "Smoke detector",
        fromZigbee: [legacy.fz.tuya_alecto_smoke],
        toZigbee: [legacy.tz.tuya_alecto_smoke],
        meta: {},
        exposes: [
            e.enum("smoke_state", ea.STATE, ["alarm", "normal"]),
            e.enum("battery_state", ea.STATE, ["low", "middle", "high"]),
            e.enum("checking_result", ea.STATE, ["checking", "check_success", "check_failure", "others"]),
            e.numeric("smoke_value", ea.STATE),
            e.numeric("battery", ea.STATE),
            e.binary("lifecycle", ea.STATE, true, false),
            e.binary("self_checking", ea.STATE_SET, true, false),
            e.binary("silence", ea.STATE_SET, true, false),
        ],
    },
];
