// zigbee-herdsman-converters/dist/devices/enocean.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.definitions = void 0;
const fz = __importStar(require("../converters/fromZigbee"));
const exposes = __importStar(require("../lib/exposes"));
const e = exposes.presets;
exports.definitions = [
    {
        fingerprint: [{ modelID: "GreenPower_2", ieeeAddr: /^0x00000000017.....$/ }],
        model: "PTM 215Z",
        vendor: "EnOcean",
        description: "Pushbutton transmitter module",
        fromZigbee: [fz.enocean_ptm215z],
        toZigbee: [],
        exposes: [
            e.action([
                "press_1",
                "release_1",
                "press_2",
                "release_2",
                "press_3",
                "release_3",
                "press_4",
                "release_4",
                "press_1_and_3",
                "release_1_and_3",
                "press_2_and_4",
                "release_2_and_4",
                "press_energy_bar",
            ]),
        ],
        whiteLabel: [
            { vendor: "Niko", description: "Dimmer switch for Hue system", model: "91004" },
            { vendor: "NodOn", description: "Smart switch for Philips Hue", model: "CWS-4-1-01_HUE" },
            { vendor: "Vimar", description: "Zigbee Friends of Hue RF switch", model: "03906" },
            { vendor: "Sunricher", model: "SR-ZGP2801K4-FOH-E" },
            { vendor: "Sunricher", model: "SR-ZG2833PAC" },
            { vendor: "LED-Trading", model: "9125" },
            { vendor: "Feller", description: "Smart light control for Philips Hue", model: "4120.2.S.FMI.61" },
            { vendor: "Namron", description: " Zigbee FOH Green Bryter K4", model: "4512727" },
        ],
    },
    {
        fingerprint: [{ modelID: "GreenPower_2", ieeeAddr: /^0x00000000015.....$/ }],
        model: "PTM 215ZE",
        vendor: "EnOcean",
        description: "Pushbutton transmitter module",
        fromZigbee: [fz.enocean_ptm215ze],
        toZigbee: [],
        exposes: [
            e.action([
                "press_1",
                "release_1",
                "press_2",
                "release_2",
                "press_3",
                "release_3",
                "press_4",
                "release_4",
                "press_1_and_2",
                "release_1_and_2",
                "press_1_and_3",
                "release_1_and_3",
                "press_1_and_4",
                "release_1_and_4",
                "press_2_and_3",
                "release_2_and_3",
                "press_2_and_4",
                "release_2_and_4",
                "press_3_and_4",
                "release_3_and_4",
                "press_energy_bar",
                "release_energy_bar",
                "press_or_release_all",
                "lock",
                "unlock",
                "half_open",
                "tilt",
            ]),
        ],
        whiteLabel: [
            { vendor: "Easyfit by EnOcean", description: "Wall switch for Zigbee", model: "EWSxZ" },
            { vendor: "Trio2sys", description: "Zigbee Green Power complete switch", model: "20020002" },
        ],
    },
    {
        fingerprint: [{ modelID: "GreenPower_7", ieeeAddr: /^0x00000000015.....$/ }],
        model: "PTM 216Z",
        vendor: "EnOcean",
        description: "Pushbutton transmitter module",
        fromZigbee: [fz.enocean_ptm216z],
        toZigbee: [],
        exposes: [
            e.action([
                "press_1",
                "press_2",
                "press_1_and_2",
                "press_3",
                "press_1_and_3",
                "press_2_and_3",
                "press_1_and_2_and_3",
                "press_4",
                "press_1_and_4",
                "press_2_and_4",
                "press_1_and_2_and_4",
                "press_3_and_4",
                "press_1_and_3_and_4",
                "press_2_and_3_and_4",
                "press_all",
                "press_energy_bar",
                "release",
                "short_press_2_of_2",
            ]),
        ],
    },
];
