// zigbee-herdsman-converters/dist/devices/frient.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.definitions = void 0;
const develco_1 = require("../lib/develco");
const m = __importStar(require("../lib/modernExtend"));
// NOTE! Develco and Frient is the same company, therefore we use develco specific things in here.
exports.definitions = [
    {
        zigbeeModel: ["WISZB-131"],
        model: "WISZB-131",
        vendor: "Frient",
        description: "Temperature and contact sensor",
        extend: [m.battery(), m.iasZoneAlarm({ zoneType: "contact", zoneAttributes: ["alarm_1", "battery_low"] }), m.temperature()],
    },
    {
        zigbeeModel: ["EMIZB-141"],
        model: "EMIZB-141",
        vendor: "Frient",
        description: "Electricity meter interface 2 LED",
        extend: [
            m.electricityMeter({ cluster: "metering", power: { divisor: 1000, multiplier: 1 }, energy: { divisor: 1000, multiplier: 1 } }),
            m.battery(),
            develco_1.develcoModernExtend.addCustomClusterManuSpecificDevelcoGenBasic(),
            develco_1.develcoModernExtend.readGenBasicPrimaryVersions(),
            develco_1.develcoModernExtend.pulseConfiguration(),
            develco_1.develcoModernExtend.currentSummation(),
        ],
        ota: true,
    },
    {
        zigbeeModel: ["SMRZB-153"],
        model: "SMRZB-153",
        vendor: "Frient",
        description: "Smart Cable - Power switch with power measurement",
        extend: [m.onOff({ configureReporting: false }), m.electricityMeter()],
        ota: true,
        endpoint: () => {
            return { default: 2 };
        },
    },
    {
        zigbeeModel: ["EMIZB-151"],
        model: "EMIZB-151",
        vendor: "Frient",
        description: "Electricity Meter Interface 2 P1",
        extend: [m.electricityMeter({ threePhase: true, producedEnergy: true, tariffs: true })],
        ota: true,
        endpoint: (device) => ({ default: 2 }),
    },
];
