// zigbee-herdsman-converters/dist/devices/lightsolutions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.definitions = void 0;
const m = __importStar(require("../lib/modernExtend"));
exports.definitions = [
    {
        zigbeeModel: ["91-947"],
        model: "200403V2-B",
        vendor: "Light Solutions",
        description: "Mini dimmer 200W",
        extend: [m.light({ configureReporting: true })],
    },
    {
        zigbeeModel: ["91-948"],
        model: "200106V3",
        vendor: "Light Solutions",
        description: "Zigbee switch 200W",
        extend: [m.onOff()],
    },
    {
        zigbeeModel: ["42-032"],
        model: "42-032",
        vendor: "Light Solutions",
        description: "LED driver CCT 12V - 30W - CCT",
        extend: [m.light({ colorTemp: { range: [160, 450] } })],
    },
    {
        fingerprint: [{ modelID: "Dimmer-Switch-ZB3.0", manufacturerName: "Light Solutions" }],
        model: "3004482/3137308/3137309",
        vendor: "Light Solutions",
        description: "Zigbee dimmer for wire",
        extend: [m.light({ configureReporting: true })],
    },
    {
        zigbeeModel: ["91-943-PRO-SINGLE"],
        model: "91-943-PRO",
        vendor: "Light Solutions",
        description: "LED Controller 12/24v",
        extend: [m.light()],
    },
    {
        zigbeeModel: ["91-943"],
        model: "91-943",
        vendor: "Light Solutions",
        description: "LED Controller 12/24v",
        extend: [m.light()],
    },
];
