// zigbee-herdsman-converters/dist/devices/skydance.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.definitions = void 0;
const exposes = __importStar(require("../lib/exposes"));
const legacy = __importStar(require("../lib/legacy"));
const tuya = __importStar(require("../lib/tuya"));
const e = exposes.presets;
const ea = exposes.access;
exports.definitions = [
    {
        fingerprint: tuya.fingerprint("TS0601", ["_TZE200_6qoazbre", "_TZE200_fcooykb4"]),
        model: "WZ5_dim_1",
        vendor: "Skydance",
        description: "Zigbee & RF 5 in 1 LED controller (DIM mode)",
        fromZigbee: [legacy.fz.tuya_light_wz5],
        toZigbee: [legacy.tz.tuya_dimmer_state, legacy.tz.tuya_light_wz5],
        exposes: [e.light().withBrightness().setAccess("state", ea.STATE_SET).setAccess("brightness", ea.STATE_SET)],
    },
    {
        fingerprint: tuya.fingerprint("TS0601", ["_TZE200_gz3n0tzf", "_TZE200_nthosjmx", "_TZE200_na98lvjp", "_TZE200_2gtsuokt"]),
        model: "WZ5_cct",
        vendor: "Skydance",
        description: "Zigbee & RF 5 in 1 LED controller (CCT mode)",
        fromZigbee: [legacy.fz.tuya_light_wz5],
        toZigbee: [legacy.tz.tuya_dimmer_state, legacy.tz.tuya_light_wz5],
        exposes: [
            e
                .light()
                .withBrightness()
                .setAccess("state", ea.STATE_SET)
                .setAccess("brightness", ea.STATE_SET)
                .withColorTemp([250, 454])
                .setAccess("color_temp", ea.STATE_SET),
        ],
        whiteLabel: [
            tuya.whitelabel("Ltech", "TY-75-24-G2Z2_CCT", "150W 24V Zigbee CV tunable white LED driver", ["_TZE200_na98lvjp"]),
            tuya.whitelabel("Ltech", "SE-20-250-1000-W2Z2", "LED (CCT mode)", ["_TZE200_2gtsuokt"]),
        ],
    },
    {
        fingerprint: tuya.fingerprint("TS0601", ["_TZE200_9hghastn", "_TZE200_9mt3kgn0"]),
        model: "WZ5_rgb",
        vendor: "Skydance",
        description: "Zigbee & RF 5 in 1 LED controller (RGB mode)",
        fromZigbee: [legacy.fz.tuya_light_wz5],
        toZigbee: [legacy.tz.tuya_dimmer_state, legacy.tz.tuya_light_wz5],
        exposes: [
            e
                .light()
                .withBrightness()
                .setAccess("state", ea.STATE_SET)
                .setAccess("brightness", ea.STATE_SET)
                .withColor(["hs"])
                .setAccess("color_hs", ea.STATE_SET),
        ],
    },
    {
        fingerprint: tuya.fingerprint("TS0503B", ["_TZB210_zdvrsts8"]),
        model: "WZ5_rgb_1",
        vendor: "Tuya",
        description: "Zigbee & RF 5 in 1 LED controller (RGB mode)",
        extend: [tuya.modernExtend.tuyaLight({ color: { modes: ["hs", "xy"] } })],
    },
    {
        fingerprint: tuya.fingerprint("TS0601", ["_TZE200_3thxjahu", "_TZE200_g9jdneiu"]),
        model: "WZ5_rgbw",
        vendor: "Skydance",
        description: "Zigbee & RF 5 in 1 LED controller (RGBW mode)",
        fromZigbee: [legacy.fz.tuya_light_wz5],
        toZigbee: [legacy.tz.tuya_dimmer_state, legacy.tz.tuya_light_wz5],
        exposes: [
            e
                .light()
                .withBrightness()
                .setAccess("state", ea.STATE_SET)
                .setAccess("brightness", ea.STATE_SET)
                .withColor(["hs"])
                .setAccess("color_hs", ea.STATE_SET),
            e.numeric("white_brightness", ea.STATE_SET).withValueMin(0).withValueMax(254).withDescription("White brightness of this light"),
        ],
        meta: { separateWhite: true },
    },
    {
        fingerprint: tuya.fingerprint("TS0601", ["_TZE200_mde0utnv", "_TZE200_aa9awrng"]),
        model: "WZ5_rgbcct",
        vendor: "Skydance",
        description: "Zigbee & RF 5 in 1 LED controller (RGB+CCT mode)",
        fromZigbee: [legacy.fz.tuya_light_wz5],
        toZigbee: [legacy.tz.tuya_dimmer_state, legacy.tz.tuya_light_wz5],
        exposes: [
            e
                .light()
                .withBrightness()
                .setAccess("state", ea.STATE_SET)
                .setAccess("brightness", ea.STATE_SET)
                .withColor(["hs"])
                .withColorTemp([250, 454])
                .setAccess("color_temp", ea.STATE_SET)
                .setAccess("color_hs", ea.STATE_SET),
            e.numeric("white_brightness", ea.STATE_SET).withValueMin(0).withValueMax(254).withDescription("White brightness of this light"),
        ],
        meta: { separateWhite: true },
    },
];
