// zigbee-herdsman-converters/dist/devices/smlight.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.definitions = void 0;
const m = __importStar(require("../lib/modernExtend"));
exports.definitions = [
    {
        zigbeeModel: [
            "SLZB-06M",
            "SLZB-06",
            "SLZB-06P7",
            "SLZB-06P10",
            "SLZB-07",
            "SLZB-07P10",
            "SLZB-07P7",
            "SLZB-0xp7",
            "SLZB-07Mg24",
            "SLZB-06Mg24",
            "SLZB-06Mg26",
            "SLZB-MR3 CC2674P10",
        ],
        model: "SLZB-06M",
        vendor: "SMLIGHT",
        description: "Router",
        whiteLabel: [
            { vendor: "SMLIGHT", model: "SLZB-06", description: "Router", fingerprint: [{ modelID: "SLZB-06" }] },
            { vendor: "SMLIGHT", model: "SLZB-06P7", description: "Router", fingerprint: [{ modelID: "SLZB-06P7" }] },
            { vendor: "SMLIGHT", model: "SLZB-06p10", description: "Router", fingerprint: [{ modelID: "SLZB-06P10" }] },
            { vendor: "SMLIGHT", model: "SLZB-07", description: "Router", fingerprint: [{ modelID: "SLZB-07" }] },
            { vendor: "SMLIGHT", model: "SLZB-07P10", description: "Router", fingerprint: [{ modelID: "SLZB-07P10" }] },
            { vendor: "SMLIGHT", model: "SLZB-07P7", description: "Router", fingerprint: [{ modelID: "SLZB-07P7" }] },
            { vendor: "SMLIGHT", model: "SLZB-0xp7", description: "Router", fingerprint: [{ modelID: "SLZB-0xp7" }] },
            { vendor: "SMLIGHT", model: "SLZB-07Mg24", description: "Router", fingerprint: [{ modelID: "SLZB-07Mg24" }] },
            { vendor: "SMLIGHT", model: "SLZB-06Mg24", description: "Router", fingerprint: [{ modelID: "SLZB-06Mg24" }] },
            { vendor: "SMLIGHT", model: "SLZB-06Mg26", description: "Router", fingerprint: [{ modelID: "SLZB-06Mg26" }] },
            { vendor: "SMLIGHT", model: "SLZB-MR3", description: "Router", fingerprint: [{ modelID: "SLZB-MR3 CC2674P10" }] },
        ],
        extend: [m.linkQuality({ reporting: true }), m.forcePowerSource({ powerSource: "Mains (single phase)" })],
    },
];
