// zigbee-herdsman/dist/zspec/zcl/definition/manufacturerCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManufacturerCode = void 0;
/**
 * https://github.com/SiliconLabs/gecko_sdk/blob/gsdk_4.4/app/zcl/manufacturers.xml
 */
var ManufacturerCode;
(function (ManufacturerCode) {
    ManufacturerCode[ManufacturerCode["MATTER_STANDARD"] = 0] = "MATTER_STANDARD";
    ManufacturerCode[ManufacturerCode["PANASONIC"] = 1] = "PANASONIC";
    ManufacturerCode[ManufacturerCode["SONY"] = 2] = "SONY";
    ManufacturerCode[ManufacturerCode["SAMSUNG"] = 3] = "SAMSUNG";
    ManufacturerCode[ManufacturerCode["PHILIPS_RF4CE"] = 4] = "PHILIPS_RF4CE";
    ManufacturerCode[ManufacturerCode["FREESCALE_RF4CE"] = 5] = "FREESCALE_RF4CE";
    ManufacturerCode[ManufacturerCode["OKI_SEMICONDUCTORS_RF4CE"] = 6] = "OKI_SEMICONDUCTORS_RF4CE";
    ManufacturerCode[ManufacturerCode["TEXAS_INSTRUMENTS"] = 7] = "TEXAS_INSTRUMENTS";
    ManufacturerCode[ManufacturerCode["CIRRONET"] = 4096] = "CIRRONET";
    ManufacturerCode[ManufacturerCode["CHIPCON"] = 4097] = "CHIPCON";
    ManufacturerCode[ManufacturerCode["EMBER"] = 4098] = "EMBER";
    ManufacturerCode[ManufacturerCode["NTS"] = 4099] = "NTS";
    ManufacturerCode[ManufacturerCode["FREESCALE"] = 4100] = "FREESCALE";
    ManufacturerCode[ManufacturerCode["IP_COM"] = 4101] = "IP_COM";
    ManufacturerCode[ManufacturerCode["SAN_JUAN_SOFTWARE"] = 4102] = "SAN_JUAN_SOFTWARE";
    ManufacturerCode[ManufacturerCode["TUV"] = 4103] = "TUV";
    ManufacturerCode[ManufacturerCode["INTEGRATION"] = 4104] = "INTEGRATION";
    ManufacturerCode[ManufacturerCode["BM_SPA"] = 4105] = "BM_SPA";
    ManufacturerCode[ManufacturerCode["AWAREPOINT"] = 4106] = "AWAREPOINT";
    ManufacturerCode[ManufacturerCode["SIGNIFY_NETHERLANDS_B_V"] = 4107] = "SIGNIFY_NETHERLANDS_B_V";
    ManufacturerCode[ManufacturerCode["LUXOFT"] = 4108] = "LUXOFT";
    ManufacturerCode[ManufacturerCode["KORWIN"] = 4109] = "KORWIN";
    ManufacturerCode[ManufacturerCode["ONE_RF_TECHNOLOGY"] = 4110] = "ONE_RF_TECHNOLOGY";
    ManufacturerCode[ManufacturerCode["SOFTWARE_TECHNOLOGIES_GROUP"] = 4111] = "SOFTWARE_TECHNOLOGIES_GROUP";
    ManufacturerCode[ManufacturerCode["TELEGESIS"] = 4112] = "TELEGESIS";
    ManufacturerCode[ManufacturerCode["VISONIC"] = 4113] = "VISONIC";
    ManufacturerCode[ManufacturerCode["INSTA"] = 4114] = "INSTA";
    ManufacturerCode[ManufacturerCode["ATALUM"] = 4115] = "ATALUM";
    ManufacturerCode[ManufacturerCode["ATMEL"] = 4116] = "ATMEL";
    ManufacturerCode[ManufacturerCode["DEVELCO"] = 4117] = "DEVELCO";
    ManufacturerCode[ManufacturerCode["HONEYWELL"] = 4118] = "HONEYWELL";
    ManufacturerCode[ManufacturerCode["RADIOPULSE"] = 4119] = "RADIOPULSE";
    ManufacturerCode[ManufacturerCode["RENESAS"] = 4120] = "RENESAS";
    ManufacturerCode[ManufacturerCode["XANADU_WIRELESS"] = 4121] = "XANADU_WIRELESS";
    ManufacturerCode[ManufacturerCode["NEC_ENGINEERING"] = 4122] = "NEC_ENGINEERING";
    ManufacturerCode[ManufacturerCode["YAMATAKE_CORPORATION"] = 4123] = "YAMATAKE_CORPORATION";
    ManufacturerCode[ManufacturerCode["TENDRIL_NETWORKS"] = 4124] = "TENDRIL_NETWORKS";
    ManufacturerCode[ManufacturerCode["ASSA_ABLOY"] = 4125] = "ASSA_ABLOY";
    ManufacturerCode[ManufacturerCode["MAXSTREAM"] = 4126] = "MAXSTREAM";
    ManufacturerCode[ManufacturerCode["NEUROCOM"] = 4127] = "NEUROCOM";
    ManufacturerCode[ManufacturerCode["INSTITUTE_FOR_INFORMATION_INDUSTRY"] = 4128] = "INSTITUTE_FOR_INFORMATION_INDUSTRY";
    ManufacturerCode[ManufacturerCode["LEGRAND_GROUP"] = 4129] = "LEGRAND_GROUP";
    ManufacturerCode[ManufacturerCode["ICONTROL"] = 4130] = "ICONTROL";
    ManufacturerCode[ManufacturerCode["RAYMARINE"] = 4131] = "RAYMARINE";
    ManufacturerCode[ManufacturerCode["LS_RESEARCH"] = 4132] = "LS_RESEARCH";
    ManufacturerCode[ManufacturerCode["ONITY_INC"] = 4133] = "ONITY_INC";
    ManufacturerCode[ManufacturerCode["MONO_PRODUCTS"] = 4134] = "MONO_PRODUCTS";
    ManufacturerCode[ManufacturerCode["RF_TECHNOLOGIES"] = 4135] = "RF_TECHNOLOGIES";
    ManufacturerCode[ManufacturerCode["ITRON"] = 4136] = "ITRON";
    ManufacturerCode[ManufacturerCode["TRITECH"] = 4137] = "TRITECH";
    ManufacturerCode[ManufacturerCode["EMBEDIT_A_S"] = 4138] = "EMBEDIT_A_S";
    ManufacturerCode[ManufacturerCode["S3C"] = 4139] = "S3C";
    ManufacturerCode[ManufacturerCode["SIEMENS"] = 4140] = "SIEMENS";
    ManufacturerCode[ManufacturerCode["MINDTECH"] = 4141] = "MINDTECH";
    ManufacturerCode[ManufacturerCode["LG_ELECTRONICS"] = 4142] = "LG_ELECTRONICS";
    ManufacturerCode[ManufacturerCode["MITSUBISHI_ELECTRIC_CORP"] = 4143] = "MITSUBISHI_ELECTRIC_CORP";
    ManufacturerCode[ManufacturerCode["JOHNSON_CONTROLS"] = 4144] = "JOHNSON_CONTROLS";
    ManufacturerCode[ManufacturerCode["SECURE_METERS_UK_LTD"] = 4145] = "SECURE_METERS_UK_LTD";
    ManufacturerCode[ManufacturerCode["KNICK"] = 4146] = "KNICK";
    ManufacturerCode[ManufacturerCode["VICONICS"] = 4147] = "VICONICS";
    ManufacturerCode[ManufacturerCode["FLEXIPANEL"] = 4148] = "FLEXIPANEL";
    ManufacturerCode[ManufacturerCode["PIASIM_CORPORATION_PTE_LTD"] = 4149] = "PIASIM_CORPORATION_PTE_LTD";
    ManufacturerCode[ManufacturerCode["TRANE"] = 4150] = "TRANE";
    ManufacturerCode[ManufacturerCode["NXP_SEMICONDUCTORS"] = 4151] = "NXP_SEMICONDUCTORS";
    ManufacturerCode[ManufacturerCode["LIVING_INDEPENDENTLY_GROUP"] = 4152] = "LIVING_INDEPENDENTLY_GROUP";
    ManufacturerCode[ManufacturerCode["ALERTME_COM"] = 4153] = "ALERTME_COM";
    ManufacturerCode[ManufacturerCode["DAINTREE"] = 4154] = "DAINTREE";
    ManufacturerCode[ManufacturerCode["AIJI_SYSTEM"] = 4155] = "AIJI_SYSTEM";
    ManufacturerCode[ManufacturerCode["TELECOM_ITALIA"] = 4156] = "TELECOM_ITALIA";
    ManufacturerCode[ManufacturerCode["MIKROKRETS_AS"] = 4157] = "MIKROKRETS_AS";
    ManufacturerCode[ManufacturerCode["OKI_SEMICONDUCTOR"] = 4158] = "OKI_SEMICONDUCTOR";
    ManufacturerCode[ManufacturerCode["NEWPORT_ELECTONICS"] = 4159] = "NEWPORT_ELECTONICS";
    ManufacturerCode[ManufacturerCode["CONTROL_4"] = 4160] = "CONTROL_4";
    ManufacturerCode[ManufacturerCode["STMICROELECTRONICS"] = 4161] = "STMICROELECTRONICS";
    ManufacturerCode[ManufacturerCode["AD_SOL_NISSIN_CORP"] = 4162] = "AD_SOL_NISSIN_CORP";
    ManufacturerCode[ManufacturerCode["DCSI"] = 4163] = "DCSI";
    ManufacturerCode[ManufacturerCode["FRANCE_TELECOM"] = 4164] = "FRANCE_TELECOM";
    ManufacturerCode[ManufacturerCode["MUNET"] = 4165] = "MUNET";
    ManufacturerCode[ManufacturerCode["AUTANI_CORPORATION"] = 4166] = "AUTANI_CORPORATION";
    ManufacturerCode[ManufacturerCode["COLORADO_VNET"] = 4167] = "COLORADO_VNET";
    ManufacturerCode[ManufacturerCode["AEROCOMM_INC"] = 4168] = "AEROCOMM_INC";
    ManufacturerCode[ManufacturerCode["SILICON_LABORATORIES"] = 4169] = "SILICON_LABORATORIES";
    ManufacturerCode[ManufacturerCode["INNCOM_INTERNATIONAL_INC"] = 4170] = "INNCOM_INTERNATIONAL_INC";
    ManufacturerCode[ManufacturerCode["COOPER_POWER_SYSTEMS"] = 4171] = "COOPER_POWER_SYSTEMS";
    ManufacturerCode[ManufacturerCode["SYNAPSE"] = 4172] = "SYNAPSE";
    ManufacturerCode[ManufacturerCode["FISHER_PIERCE_SUNRISE"] = 4173] = "FISHER_PIERCE_SUNRISE";
    ManufacturerCode[ManufacturerCode["CENTRALITE_SYSTEMS_INC"] = 4174] = "CENTRALITE_SYSTEMS_INC";
    ManufacturerCode[ManufacturerCode["CRANE_WIRELESS_MONITORING_SOLUTIONS"] = 4175] = "CRANE_WIRELESS_MONITORING_SOLUTIONS";
    ManufacturerCode[ManufacturerCode["MOBILARM_LIMITED"] = 4176] = "MOBILARM_LIMITED";
    ManufacturerCode[ManufacturerCode["IMONITOR_RESEARCH_LTD"] = 4177] = "IMONITOR_RESEARCH_LTD";
    ManufacturerCode[ManufacturerCode["BARTECH"] = 4178] = "BARTECH";
    ManufacturerCode[ManufacturerCode["MESHNETICS"] = 4179] = "MESHNETICS";
    ManufacturerCode[ManufacturerCode["LS_INDUSTRIAL_SYSTEMS_CO_LTD"] = 4180] = "LS_INDUSTRIAL_SYSTEMS_CO_LTD";
    ManufacturerCode[ManufacturerCode["CASON_ENGINEERING_PLC"] = 4181] = "CASON_ENGINEERING_PLC";
    ManufacturerCode[ManufacturerCode["WIRELESS_GLUE_NETWORKS_INC"] = 4182] = "WIRELESS_GLUE_NETWORKS_INC";
    ManufacturerCode[ManufacturerCode["ELSTER"] = 4183] = "ELSTER";
    ManufacturerCode[ManufacturerCode["SMS_TECNOLOGIA_ELETRONICA"] = 4184] = "SMS_TECNOLOGIA_ELETRONICA";
    ManufacturerCode[ManufacturerCode["ONSET_COMPUTER_CORPORATION"] = 4185] = "ONSET_COMPUTER_CORPORATION";
    ManufacturerCode[ManufacturerCode["RIGA_DEVELOPMENT"] = 4186] = "RIGA_DEVELOPMENT";
    ManufacturerCode[ManufacturerCode["ENERGATE"] = 4187] = "ENERGATE";
    ManufacturerCode[ManufacturerCode["CONMED_LINVATEC"] = 4188] = "CONMED_LINVATEC";
    ManufacturerCode[ManufacturerCode["POWERMAND"] = 4189] = "POWERMAND";
    ManufacturerCode[ManufacturerCode["SCHNEIDER_ELECTRIC"] = 4190] = "SCHNEIDER_ELECTRIC";
    ManufacturerCode[ManufacturerCode["EATON_CORPORATION"] = 4191] = "EATON_CORPORATION";
    ManufacturerCode[ManufacturerCode["TELULAR_CORPORATION"] = 4192] = "TELULAR_CORPORATION";
    ManufacturerCode[ManufacturerCode["DELPHI_MEDICAL_SYSTEMS"] = 4193] = "DELPHI_MEDICAL_SYSTEMS";
    ManufacturerCode[ManufacturerCode["EPISENSOR_LIMITED"] = 4194] = "EPISENSOR_LIMITED";
    ManufacturerCode[ManufacturerCode["LANDIS_GYR"] = 4195] = "LANDIS_GYR";
    ManufacturerCode[ManufacturerCode["KABA_GROUP"] = 4196] = "KABA_GROUP";
    ManufacturerCode[ManufacturerCode["SHURE_INCORPORATED"] = 4197] = "SHURE_INCORPORATED";
    ManufacturerCode[ManufacturerCode["COMVERGE_INC"] = 4198] = "COMVERGE_INC";
    ManufacturerCode[ManufacturerCode["DBS_LODGING_TECHNOLOGIES_LLC"] = 4199] = "DBS_LODGING_TECHNOLOGIES_LLC";
    ManufacturerCode[ManufacturerCode["ENERGY_AWARE_TECHNOLOGY_INC"] = 4200] = "ENERGY_AWARE_TECHNOLOGY_INC";
    ManufacturerCode[ManufacturerCode["HIDALGO_LIMITED"] = 4201] = "HIDALGO_LIMITED";
    ManufacturerCode[ManufacturerCode["AIR2APP"] = 4202] = "AIR2APP";
    ManufacturerCode[ManufacturerCode["AMX"] = 4203] = "AMX";
    ManufacturerCode[ManufacturerCode["EDMI_PTY_LTD"] = 4204] = "EDMI_PTY_LTD";
    ManufacturerCode[ManufacturerCode["CYAN_LTD"] = 4205] = "CYAN_LTD";
    ManufacturerCode[ManufacturerCode["SYSTEM_SPA"] = 4206] = "SYSTEM_SPA";
    ManufacturerCode[ManufacturerCode["TELIT"] = 4207] = "TELIT";
    ManufacturerCode[ManufacturerCode["KAGA_ELECTRONICS"] = 4208] = "KAGA_ELECTRONICS";
    ManufacturerCode[ManufacturerCode["ASTREL_GROUP_SRL"] = 4209] = "ASTREL_GROUP_SRL";
    ManufacturerCode[ManufacturerCode["CERTICOM"] = 4210] = "CERTICOM";
    ManufacturerCode[ManufacturerCode["GRIDPOINT"] = 4211] = "GRIDPOINT";
    ManufacturerCode[ManufacturerCode["PROFILE_SYSTEMS_LLC"] = 4212] = "PROFILE_SYSTEMS_LLC";
    ManufacturerCode[ManufacturerCode["COMPACTA_INTERNATIONAL_LTD"] = 4213] = "COMPACTA_INTERNATIONAL_LTD";
    ManufacturerCode[ManufacturerCode["FREESTYLE_TECHNOLOGY_PTY_LTD"] = 4214] = "FREESTYLE_TECHNOLOGY_PTY_LTD";
    ManufacturerCode[ManufacturerCode["ALEKTRONA"] = 4215] = "ALEKTRONA";
    ManufacturerCode[ManufacturerCode["COMPUTIME"] = 4216] = "COMPUTIME";
    ManufacturerCode[ManufacturerCode["REMOTE_TECHNOLOGIES_INC"] = 4217] = "REMOTE_TECHNOLOGIES_INC";
    ManufacturerCode[ManufacturerCode["WAVECOM_S_A"] = 4218] = "WAVECOM_S_A";
    ManufacturerCode[ManufacturerCode["ENERGY_OPTIMIZERS_LTD"] = 4219] = "ENERGY_OPTIMIZERS_LTD";
    ManufacturerCode[ManufacturerCode["GE"] = 4220] = "GE";
    ManufacturerCode[ManufacturerCode["JETLUN"] = 4221] = "JETLUN";
    ManufacturerCode[ManufacturerCode["CIPHER_SYSTEMS"] = 4222] = "CIPHER_SYSTEMS";
    ManufacturerCode[ManufacturerCode["CORPORATE_SYSTEMS_ENGINEERING"] = 4223] = "CORPORATE_SYSTEMS_ENGINEERING";
    ManufacturerCode[ManufacturerCode["ECOBEE"] = 4224] = "ECOBEE";
    ManufacturerCode[ManufacturerCode["SMK"] = 4225] = "SMK";
    ManufacturerCode[ManufacturerCode["MESHWORKS_WIRELESS_OY"] = 4226] = "MESHWORKS_WIRELESS_OY";
    ManufacturerCode[ManufacturerCode["ELLIPS_B_V"] = 4227] = "ELLIPS_B_V";
    ManufacturerCode[ManufacturerCode["SECURE_ELECTRANS"] = 4228] = "SECURE_ELECTRANS";
    ManufacturerCode[ManufacturerCode["CEDO"] = 4229] = "CEDO";
    ManufacturerCode[ManufacturerCode["TOSHIBA"] = 4230] = "TOSHIBA";
    ManufacturerCode[ManufacturerCode["DIGI_INTERNATIONAL"] = 4231] = "DIGI_INTERNATIONAL";
    ManufacturerCode[ManufacturerCode["UBILOGIX"] = 4232] = "UBILOGIX";
    ManufacturerCode[ManufacturerCode["ECHELON"] = 4233] = "ECHELON";
    ManufacturerCode[ManufacturerCode["GREEN_ENERGY_OPTIONS"] = 4240] = "GREEN_ENERGY_OPTIONS";
    ManufacturerCode[ManufacturerCode["SILVER_SPRING_NETWORKS"] = 4241] = "SILVER_SPRING_NETWORKS";
    ManufacturerCode[ManufacturerCode["BLACK_AND_DECKER"] = 4242] = "BLACK_AND_DECKER";
    ManufacturerCode[ManufacturerCode["AZTECH_ASSOCIATES_INC"] = 4243] = "AZTECH_ASSOCIATES_INC";
    ManufacturerCode[ManufacturerCode["AANDD_CO_LTD"] = 4244] = "AANDD_CO_LTD";
    ManufacturerCode[ManufacturerCode["RAINFOREST_AUTOMATION"] = 4245] = "RAINFOREST_AUTOMATION";
    ManufacturerCode[ManufacturerCode["CARRIER_ELECTRONICS"] = 4246] = "CARRIER_ELECTRONICS";
    ManufacturerCode[ManufacturerCode["SYCHIP_MURATA"] = 4247] = "SYCHIP_MURATA";
    ManufacturerCode[ManufacturerCode["OPENPEAK"] = 4248] = "OPENPEAK";
    ManufacturerCode[ManufacturerCode["PASSIVESYSTEMS"] = 4249] = "PASSIVESYSTEMS";
    ManufacturerCode[ManufacturerCode["MMB_RESEARCH"] = 4250] = "MMB_RESEARCH";
    ManufacturerCode[ManufacturerCode["LEVITON_MANUFACTURING_COMPANY"] = 4251] = "LEVITON_MANUFACTURING_COMPANY";
    ManufacturerCode[ManufacturerCode["KOREA_ELECTRIC_POWER_DATA_NETWORK_CO_LTD"] = 4252] = "KOREA_ELECTRIC_POWER_DATA_NETWORK_CO_LTD";
    ManufacturerCode[ManufacturerCode["COMCAST"] = 4253] = "COMCAST";
    ManufacturerCode[ManufacturerCode["NEC_ELECTRONICS"] = 4254] = "NEC_ELECTRONICS";
    ManufacturerCode[ManufacturerCode["NETVOX"] = 4255] = "NETVOX";
    ManufacturerCode[ManufacturerCode["U_CONTROL"] = 4256] = "U_CONTROL";
    ManufacturerCode[ManufacturerCode["EMBEDIA_TECHNOLOGIES_CORP"] = 4257] = "EMBEDIA_TECHNOLOGIES_CORP";
    ManufacturerCode[ManufacturerCode["SENSUS"] = 4258] = "SENSUS";
    ManufacturerCode[ManufacturerCode["SUNRISE_TECHNOLOGIES"] = 4259] = "SUNRISE_TECHNOLOGIES";
    ManufacturerCode[ManufacturerCode["MEMTECH_CORP"] = 4260] = "MEMTECH_CORP";
    ManufacturerCode[ManufacturerCode["FREEBOX"] = 4261] = "FREEBOX";
    ManufacturerCode[ManufacturerCode["M2_LABS_LTD"] = 4262] = "M2_LABS_LTD";
    ManufacturerCode[ManufacturerCode["BRITISH_GAS"] = 4263] = "BRITISH_GAS";
    ManufacturerCode[ManufacturerCode["SENTEC_LTD"] = 4264] = "SENTEC_LTD";
    ManufacturerCode[ManufacturerCode["NAVETAS"] = 4265] = "NAVETAS";
    ManufacturerCode[ManufacturerCode["LIGHTSPEED_TECHNOLOGIES"] = 4266] = "LIGHTSPEED_TECHNOLOGIES";
    ManufacturerCode[ManufacturerCode["OKI_ELECTRIC_INDUSTRY_CO_LTD"] = 4267] = "OKI_ELECTRIC_INDUSTRY_CO_LTD";
    ManufacturerCode[ManufacturerCode["S_I_SISTEMAS_INTELIGENTES_ELETRONICOS_LTDA"] = 4268] = "S_I_SISTEMAS_INTELIGENTES_ELETRONICOS_LTDA";
    ManufacturerCode[ManufacturerCode["DOMETIC"] = 4269] = "DOMETIC";
    ManufacturerCode[ManufacturerCode["ALPS"] = 4270] = "ALPS";
    ManufacturerCode[ManufacturerCode["ENERGYHUB"] = 4271] = "ENERGYHUB";
    ManufacturerCode[ManufacturerCode["KAMSTRUP"] = 4272] = "KAMSTRUP";
    ManufacturerCode[ManufacturerCode["ECHOSTAR"] = 4273] = "ECHOSTAR";
    ManufacturerCode[ManufacturerCode["ENERNOC"] = 4274] = "ENERNOC";
    ManufacturerCode[ManufacturerCode["ELTAV"] = 4275] = "ELTAV";
    ManufacturerCode[ManufacturerCode["BELKIN"] = 4276] = "BELKIN";
    ManufacturerCode[ManufacturerCode["XSTREAMHD_WIRELESS_VENTURES"] = 4277] = "XSTREAMHD_WIRELESS_VENTURES";
    ManufacturerCode[ManufacturerCode["SATURN_SOUTH_PTY_LTD"] = 4278] = "SATURN_SOUTH_PTY_LTD";
    ManufacturerCode[ManufacturerCode["GREENTRAPONLINE_A_S"] = 4279] = "GREENTRAPONLINE_A_S";
    ManufacturerCode[ManufacturerCode["SMARTSYNCH_INC"] = 4280] = "SMARTSYNCH_INC";
    ManufacturerCode[ManufacturerCode["NYCE_CONTROL_INC"] = 4281] = "NYCE_CONTROL_INC";
    ManufacturerCode[ManufacturerCode["ICM_CONTROLS_CORP"] = 4282] = "ICM_CONTROLS_CORP";
    ManufacturerCode[ManufacturerCode["MILLENNIUM_ELECTRONICS_PTY_LTD"] = 4283] = "MILLENNIUM_ELECTRONICS_PTY_LTD";
    ManufacturerCode[ManufacturerCode["MOTOROLA_INC"] = 4284] = "MOTOROLA_INC";
    ManufacturerCode[ManufacturerCode["EMERSON_WHITE_RODGERS"] = 4285] = "EMERSON_WHITE_RODGERS";
    ManufacturerCode[ManufacturerCode["RADIO_THERMOSTAT_COMPANY_OF_AMERICA"] = 4286] = "RADIO_THERMOSTAT_COMPANY_OF_AMERICA";
    ManufacturerCode[ManufacturerCode["OMRON_CORPORATION"] = 4287] = "OMRON_CORPORATION";
    ManufacturerCode[ManufacturerCode["GIINII_GLOBAL_LIMITED"] = 4288] = "GIINII_GLOBAL_LIMITED";
    ManufacturerCode[ManufacturerCode["FUJITSU_GENERAL_LIMITED"] = 4289] = "FUJITSU_GENERAL_LIMITED";
    ManufacturerCode[ManufacturerCode["PEEL_TECHNOLOGIES_INC"] = 4290] = "PEEL_TECHNOLOGIES_INC";
    ManufacturerCode[ManufacturerCode["ACCENT_S_P_A"] = 4291] = "ACCENT_S_P_A";
    ManufacturerCode[ManufacturerCode["BYTESNAP_DESIGN_LTD"] = 4292] = "BYTESNAP_DESIGN_LTD";
    ManufacturerCode[ManufacturerCode["NEC_TOKIN_CORPORATION"] = 4293] = "NEC_TOKIN_CORPORATION";
    ManufacturerCode[ManufacturerCode["G4S_JUSTICE_SERVICES"] = 4294] = "G4S_JUSTICE_SERVICES";
    ManufacturerCode[ManufacturerCode["TRILLIANT_NETWORKS_INC"] = 4295] = "TRILLIANT_NETWORKS_INC";
    ManufacturerCode[ManufacturerCode["ELECTROLUX_ITALIA_S_P_A"] = 4296] = "ELECTROLUX_ITALIA_S_P_A";
    ManufacturerCode[ManufacturerCode["ONZO_LTD"] = 4297] = "ONZO_LTD";
    ManufacturerCode[ManufacturerCode["ENTEK_SYSTEMS"] = 4298] = "ENTEK_SYSTEMS";
    ManufacturerCode[ManufacturerCode["PHILIPS"] = 4299] = "PHILIPS";
    ManufacturerCode[ManufacturerCode["MAINSTREAM_ENGINEERING"] = 4300] = "MAINSTREAM_ENGINEERING";
    ManufacturerCode[ManufacturerCode["INDESIT_COMPANY"] = 4301] = "INDESIT_COMPANY";
    ManufacturerCode[ManufacturerCode["THINKECO_INC"] = 4302] = "THINKECO_INC";
    ManufacturerCode[ManufacturerCode["TWODTWOC_INC"] = 4303] = "TWODTWOC_INC";
    ManufacturerCode[ManufacturerCode["QORVO"] = 4304] = "QORVO";
    ManufacturerCode[ManufacturerCode["INTERCEL"] = 4305] = "INTERCEL";
    ManufacturerCode[ManufacturerCode["LG_ELECTRONICS_2"] = 4306] = "LG_ELECTRONICS_2";
    ManufacturerCode[ManufacturerCode["MITSUMI_ELECTRIC_CO_LTD"] = 4307] = "MITSUMI_ELECTRIC_CO_LTD";
    ManufacturerCode[ManufacturerCode["MITSUMI_ELECTRIC_CO_LTD_2"] = 4308] = "MITSUMI_ELECTRIC_CO_LTD_2";
    ManufacturerCode[ManufacturerCode["ZENTRUM_MIKROELEKTRONIK_DRESDEN_AG_ZMDI"] = 4309] = "ZENTRUM_MIKROELEKTRONIK_DRESDEN_AG_ZMDI";
    ManufacturerCode[ManufacturerCode["NEST_LABS_INC"] = 4310] = "NEST_LABS_INC";
    ManufacturerCode[ManufacturerCode["EXEGIN_TECHNOLOGIES_LTD"] = 4311] = "EXEGIN_TECHNOLOGIES_LTD";
    ManufacturerCode[ManufacturerCode["HONEYWELL_2"] = 4312] = "HONEYWELL_2";
    ManufacturerCode[ManufacturerCode["TAKAHATA_PRECISION_CO_LTD"] = 4313] = "TAKAHATA_PRECISION_CO_LTD";
    ManufacturerCode[ManufacturerCode["SUMITOMO_ELECTRIC_NETWORKS_INC"] = 4314] = "SUMITOMO_ELECTRIC_NETWORKS_INC";
    ManufacturerCode[ManufacturerCode["GE_ENERGY"] = 4315] = "GE_ENERGY";
    ManufacturerCode[ManufacturerCode["GE_APPLIANCES"] = 4316] = "GE_APPLIANCES";
    ManufacturerCode[ManufacturerCode["RADIOCRAFTS_AS"] = 4317] = "RADIOCRAFTS_AS";
    ManufacturerCode[ManufacturerCode["CEIVA"] = 4318] = "CEIVA";
    ManufacturerCode[ManufacturerCode["TECANDCO_CO_LTD"] = 4319] = "TECANDCO_CO_LTD";
    ManufacturerCode[ManufacturerCode["CHAMELEON_TECHNOLOGY_UK_LTD"] = 4320] = "CHAMELEON_TECHNOLOGY_UK_LTD";
    ManufacturerCode[ManufacturerCode["SAMSUNG_2"] = 4321] = "SAMSUNG_2";
    ManufacturerCode[ManufacturerCode["RUWIDO_AUSTRIA_GMBH"] = 4322] = "RUWIDO_AUSTRIA_GMBH";
    ManufacturerCode[ManufacturerCode["HUAWEI_TECHNOLOGIES_CO_LTD"] = 4323] = "HUAWEI_TECHNOLOGIES_CO_LTD";
    ManufacturerCode[ManufacturerCode["HUAWEI_TECHNOLOGIES_CO_LTD_2"] = 4324] = "HUAWEI_TECHNOLOGIES_CO_LTD_2";
    ManufacturerCode[ManufacturerCode["GREENWAVE_REALITY"] = 4325] = "GREENWAVE_REALITY";
    ManufacturerCode[ManufacturerCode["BGLOBAL_METERING_LTD"] = 4326] = "BGLOBAL_METERING_LTD";
    ManufacturerCode[ManufacturerCode["MINDTECK"] = 4327] = "MINDTECK";
    ManufacturerCode[ManufacturerCode["INGERSOLL_RAND"] = 4328] = "INGERSOLL_RAND";
    ManufacturerCode[ManufacturerCode["DIUS_COMPUTING_PTY_LTD"] = 4329] = "DIUS_COMPUTING_PTY_LTD";
    ManufacturerCode[ManufacturerCode["EMBEDDED_AUTOMATION_INC"] = 4330] = "EMBEDDED_AUTOMATION_INC";
    ManufacturerCode[ManufacturerCode["ABB"] = 4331] = "ABB";
    ManufacturerCode[ManufacturerCode["SONY_2"] = 4332] = "SONY_2";
    ManufacturerCode[ManufacturerCode["GENUS_POWER_INFRASTRUCTURES_LIMITED"] = 4333] = "GENUS_POWER_INFRASTRUCTURES_LIMITED";
    ManufacturerCode[ManufacturerCode["UNIVERSAL_ELECTRONICS_INC"] = 4334] = "UNIVERSAL_ELECTRONICS_INC";
    ManufacturerCode[ManufacturerCode["UNIVERSAL_ELECTRONICS_INC_2"] = 4335] = "UNIVERSAL_ELECTRONICS_INC_2";
    ManufacturerCode[ManufacturerCode["METRUM_TECHNOLOGIES_LLC"] = 4336] = "METRUM_TECHNOLOGIES_LLC";
    ManufacturerCode[ManufacturerCode["CISCO"] = 4337] = "CISCO";
    ManufacturerCode[ManufacturerCode["UBISYS_TECHNOLOGIES_GMBH"] = 4338] = "UBISYS_TECHNOLOGIES_GMBH";
    ManufacturerCode[ManufacturerCode["CONSERT"] = 4339] = "CONSERT";
    ManufacturerCode[ManufacturerCode["CRESTRON_ELECTRONICS"] = 4340] = "CRESTRON_ELECTRONICS";
    ManufacturerCode[ManufacturerCode["ENPHASE_ENERGY"] = 4341] = "ENPHASE_ENERGY";
    ManufacturerCode[ManufacturerCode["INVENSYS_CONTROLS"] = 4342] = "INVENSYS_CONTROLS";
    ManufacturerCode[ManufacturerCode["MUELLER_SYSTEMS_LLC"] = 4343] = "MUELLER_SYSTEMS_LLC";
    ManufacturerCode[ManufacturerCode["AAC_TECHNOLOGIES_HOLDING"] = 4344] = "AAC_TECHNOLOGIES_HOLDING";
    ManufacturerCode[ManufacturerCode["U_NEXT_CO_LTD"] = 4345] = "U_NEXT_CO_LTD";
    ManufacturerCode[ManufacturerCode["STEELCASE_INC"] = 4346] = "STEELCASE_INC";
    ManufacturerCode[ManufacturerCode["TELEMATICS_WIRELESS"] = 4347] = "TELEMATICS_WIRELESS";
    ManufacturerCode[ManufacturerCode["SAMIL_POWER_CO_LTD"] = 4348] = "SAMIL_POWER_CO_LTD";
    ManufacturerCode[ManufacturerCode["PACE_PLC"] = 4349] = "PACE_PLC";
    ManufacturerCode[ManufacturerCode["OSBORNE_COINAGE_CO"] = 4350] = "OSBORNE_COINAGE_CO";
    ManufacturerCode[ManufacturerCode["POWERWATCH"] = 4351] = "POWERWATCH";
    ManufacturerCode[ManufacturerCode["CANDELED_GMBH"] = 4352] = "CANDELED_GMBH";
    ManufacturerCode[ManufacturerCode["FLEXGRID_S_R_L"] = 4353] = "FLEXGRID_S_R_L";
    ManufacturerCode[ManufacturerCode["HUMAX"] = 4354] = "HUMAX";
    ManufacturerCode[ManufacturerCode["UNIVERSAL_DEVICES"] = 4355] = "UNIVERSAL_DEVICES";
    ManufacturerCode[ManufacturerCode["ADVANCED_ENERGY"] = 4356] = "ADVANCED_ENERGY";
    ManufacturerCode[ManufacturerCode["BEGA_GANTENBRINK_LEUCHTEN"] = 4357] = "BEGA_GANTENBRINK_LEUCHTEN";
    ManufacturerCode[ManufacturerCode["BRUNEL_UNIVERSITY"] = 4358] = "BRUNEL_UNIVERSITY";
    ManufacturerCode[ManufacturerCode["PANASONIC_RANDD_CENTER_SINGAPORE"] = 4359] = "PANASONIC_RANDD_CENTER_SINGAPORE";
    ManufacturerCode[ManufacturerCode["ESYSTEMS_RESEARCH"] = 4360] = "ESYSTEMS_RESEARCH";
    ManufacturerCode[ManufacturerCode["PANAMAX"] = 4361] = "PANAMAX";
    ManufacturerCode[ManufacturerCode["SMARTTHINGS_INC"] = 4362] = "SMARTTHINGS_INC";
    ManufacturerCode[ManufacturerCode["EM_LITE_LTD"] = 4363] = "EM_LITE_LTD";
    ManufacturerCode[ManufacturerCode["OSRAM_SYLVANIA"] = 4364] = "OSRAM_SYLVANIA";
    ManufacturerCode[ManufacturerCode["TWO_SAVE_ENERGY_LTD"] = 4365] = "TWO_SAVE_ENERGY_LTD";
    ManufacturerCode[ManufacturerCode["PLANET_INNOVATION_PRODUCTS_PTY_LTD"] = 4366] = "PLANET_INNOVATION_PRODUCTS_PTY_LTD";
    ManufacturerCode[ManufacturerCode["AMBIENT_DEVICES_INC"] = 4367] = "AMBIENT_DEVICES_INC";
    ManufacturerCode[ManufacturerCode["PROFALUX"] = 4368] = "PROFALUX";
    ManufacturerCode[ManufacturerCode["BILLION_ELECTRIC_COMPANY_BEC"] = 4369] = "BILLION_ELECTRIC_COMPANY_BEC";
    ManufacturerCode[ManufacturerCode["EMBERTEC_PTY_LTD"] = 4370] = "EMBERTEC_PTY_LTD";
    ManufacturerCode[ManufacturerCode["IT_WATCHDOGS"] = 4371] = "IT_WATCHDOGS";
    ManufacturerCode[ManufacturerCode["RELOC"] = 4372] = "RELOC";
    ManufacturerCode[ManufacturerCode["INTEL_CORPORATION"] = 4373] = "INTEL_CORPORATION";
    ManufacturerCode[ManufacturerCode["TREND_ELECTRONICS_LIMITED"] = 4374] = "TREND_ELECTRONICS_LIMITED";
    ManufacturerCode[ManufacturerCode["MOXA"] = 4375] = "MOXA";
    ManufacturerCode[ManufacturerCode["QEES"] = 4376] = "QEES";
    ManufacturerCode[ManufacturerCode["SAYME_WIRELESS_SENSOR_NETWORKS"] = 4377] = "SAYME_WIRELESS_SENSOR_NETWORKS";
    ManufacturerCode[ManufacturerCode["PENTAIR_AQUATIC_SYSTEMS"] = 4378] = "PENTAIR_AQUATIC_SYSTEMS";
    ManufacturerCode[ManufacturerCode["ORBIT_IRRIGATION"] = 4379] = "ORBIT_IRRIGATION";
    ManufacturerCode[ManufacturerCode["CALIFORNIA_EASTERN_LABORATORIES"] = 4380] = "CALIFORNIA_EASTERN_LABORATORIES";
    ManufacturerCode[ManufacturerCode["COMCAST_2"] = 4381] = "COMCAST_2";
    ManufacturerCode[ManufacturerCode["IDT_TECHNOLOGY_LIMITED"] = 4382] = "IDT_TECHNOLOGY_LIMITED";
    ManufacturerCode[ManufacturerCode["PIXELA_CORPORATION"] = 4383] = "PIXELA_CORPORATION";
    ManufacturerCode[ManufacturerCode["TIVO_INC"] = 4384] = "TIVO_INC";
    ManufacturerCode[ManufacturerCode["FIDURE_CORP"] = 4385] = "FIDURE_CORP";
    ManufacturerCode[ManufacturerCode["MARVELL_SEMICONDUCTOR_INC"] = 4386] = "MARVELL_SEMICONDUCTOR_INC";
    ManufacturerCode[ManufacturerCode["WASION_GROUP_LIMITED"] = 4387] = "WASION_GROUP_LIMITED";
    ManufacturerCode[ManufacturerCode["JASCO_PRODUCTS_COMPANY"] = 4388] = "JASCO_PRODUCTS_COMPANY";
    ManufacturerCode[ManufacturerCode["SHENZHEN_KAIFA_TECHNOLOGY_CHENGDU_CO_LTD"] = 4389] = "SHENZHEN_KAIFA_TECHNOLOGY_CHENGDU_CO_LTD";
    ManufacturerCode[ManufacturerCode["NETCOMM_WIRELESS_LIMITED"] = 4390] = "NETCOMM_WIRELESS_LIMITED";
    ManufacturerCode[ManufacturerCode["DEFINE_INSTRUMENTS_LIMITED"] = 4391] = "DEFINE_INSTRUMENTS_LIMITED";
    ManufacturerCode[ManufacturerCode["IN_HOME_DISPLAYS_LTD"] = 4392] = "IN_HOME_DISPLAYS_LTD";
    ManufacturerCode[ManufacturerCode["MIELE_AND_CIE_KG"] = 4393] = "MIELE_AND_CIE_KG";
    ManufacturerCode[ManufacturerCode["TELEVES_S_A"] = 4394] = "TELEVES_S_A";
    ManufacturerCode[ManufacturerCode["LABELEC"] = 4395] = "LABELEC";
    ManufacturerCode[ManufacturerCode["CHINA_ELECTRONICS_STANDARDIZATION_INSTITUTE"] = 4396] = "CHINA_ELECTRONICS_STANDARDIZATION_INSTITUTE";
    ManufacturerCode[ManufacturerCode["VECTORFORM_LLC"] = 4397] = "VECTORFORM_LLC";
    ManufacturerCode[ManufacturerCode["BUSCH_JAEGER_ELEKTRO"] = 4398] = "BUSCH_JAEGER_ELEKTRO";
    ManufacturerCode[ManufacturerCode["REDPINE_SIGNALS_INC"] = 4399] = "REDPINE_SIGNALS_INC";
    ManufacturerCode[ManufacturerCode["BRIDGES_ELECTRONIC_TECHNOLOGY_PTY_LTD"] = 4400] = "BRIDGES_ELECTRONIC_TECHNOLOGY_PTY_LTD";
    ManufacturerCode[ManufacturerCode["SERCOMM"] = 4401] = "SERCOMM";
    ManufacturerCode[ManufacturerCode["WSH_GMBH_WIRSINDHELLER"] = 4402] = "WSH_GMBH_WIRSINDHELLER";
    ManufacturerCode[ManufacturerCode["BOSCH_SECURITY_SYSTEMS_INC"] = 4403] = "BOSCH_SECURITY_SYSTEMS_INC";
    ManufacturerCode[ManufacturerCode["EZEX_CORPORATION"] = 4404] = "EZEX_CORPORATION";
    ManufacturerCode[ManufacturerCode["DRESDEN_ELEKTRONIK_INGENIEURTECHNIK_GMBH"] = 4405] = "DRESDEN_ELEKTRONIK_INGENIEURTECHNIK_GMBH";
    ManufacturerCode[ManufacturerCode["MEAZON_S_A"] = 4406] = "MEAZON_S_A";
    ManufacturerCode[ManufacturerCode["CROW_ELECTRONIC_ENGINEERING_LTD"] = 4407] = "CROW_ELECTRONIC_ENGINEERING_LTD";
    ManufacturerCode[ManufacturerCode["HARVARD_ENGINEERING_PLC"] = 4408] = "HARVARD_ENGINEERING_PLC";
    ManufacturerCode[ManufacturerCode["ANDSON_BEIJING_TECHNOLOGY_CO_LTD"] = 4409] = "ANDSON_BEIJING_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["ADHOCO_AG"] = 4410] = "ADHOCO_AG";
    ManufacturerCode[ManufacturerCode["WAXMAN_CONSUMER_PRODUCTS_GROUP_INC"] = 4411] = "WAXMAN_CONSUMER_PRODUCTS_GROUP_INC";
    ManufacturerCode[ManufacturerCode["OWON_TECHNOLOGY_INC"] = 4412] = "OWON_TECHNOLOGY_INC";
    ManufacturerCode[ManufacturerCode["HITRON_TECHNOLOGIES_INC"] = 4413] = "HITRON_TECHNOLOGIES_INC";
    ManufacturerCode[ManufacturerCode["SCEMTEC_HARD_UND_SOFTWARE_FUR_MESS_UND_STEUERUNGSTECHNIK_GMBH"] = 4414] = "SCEMTEC_HARD_UND_SOFTWARE_FUR_MESS_UND_STEUERUNGSTECHNIK_GMBH";
    ManufacturerCode[ManufacturerCode["WEBEE_LLC"] = 4415] = "WEBEE_LLC";
    ManufacturerCode[ManufacturerCode["GRID2HOME_INC"] = 4416] = "GRID2HOME_INC";
    ManufacturerCode[ManufacturerCode["TELINK_MICRO"] = 4417] = "TELINK_MICRO";
    ManufacturerCode[ManufacturerCode["JASMINE_SYSTEMS_INC"] = 4418] = "JASMINE_SYSTEMS_INC";
    ManufacturerCode[ManufacturerCode["BIDGELY"] = 4419] = "BIDGELY";
    ManufacturerCode[ManufacturerCode["LUTRON"] = 4420] = "LUTRON";
    ManufacturerCode[ManufacturerCode["IJENKO"] = 4421] = "IJENKO";
    ManufacturerCode[ManufacturerCode["STARFIELD_ELECTRONIC_LTD"] = 4422] = "STARFIELD_ELECTRONIC_LTD";
    ManufacturerCode[ManufacturerCode["TCP_INC"] = 4423] = "TCP_INC";
    ManufacturerCode[ManufacturerCode["ROGERS_COMMUNICATIONS_PARTNERSHIP"] = 4424] = "ROGERS_COMMUNICATIONS_PARTNERSHIP";
    ManufacturerCode[ManufacturerCode["CREE_INC"] = 4425] = "CREE_INC";
    ManufacturerCode[ManufacturerCode["ROBERT_BOSCH_LLC"] = 4426] = "ROBERT_BOSCH_LLC";
    ManufacturerCode[ManufacturerCode["IBIS_NETWORKS_INC"] = 4427] = "IBIS_NETWORKS_INC";
    ManufacturerCode[ManufacturerCode["QUIRKY_INC"] = 4428] = "QUIRKY_INC";
    ManufacturerCode[ManufacturerCode["EFERGY_TECHNOLOGIES_LIMITED"] = 4429] = "EFERGY_TECHNOLOGIES_LIMITED";
    ManufacturerCode[ManufacturerCode["SMARTLABS_INC"] = 4430] = "SMARTLABS_INC";
    ManufacturerCode[ManufacturerCode["EVERSPRING_INDUSTRY_CO_LTD"] = 4431] = "EVERSPRING_INDUSTRY_CO_LTD";
    ManufacturerCode[ManufacturerCode["SWANN_COMMUNICATIONS_PTL_LTD"] = 4432] = "SWANN_COMMUNICATIONS_PTL_LTD";
    ManufacturerCode[ManufacturerCode["SONETER"] = 4433] = "SONETER";
    ManufacturerCode[ManufacturerCode["SAMSUNG_SDS"] = 4434] = "SAMSUNG_SDS";
    ManufacturerCode[ManufacturerCode["UNIBAND_ELECTRONIC_CORPORATION"] = 4435] = "UNIBAND_ELECTRONIC_CORPORATION";
    ManufacturerCode[ManufacturerCode["ACCTON_TECHNOLOGY_CORPORATION"] = 4436] = "ACCTON_TECHNOLOGY_CORPORATION";
    ManufacturerCode[ManufacturerCode["BOSCH_THERMOTECHNIK_GMBH"] = 4437] = "BOSCH_THERMOTECHNIK_GMBH";
    ManufacturerCode[ManufacturerCode["WINCOR_NIXDORF_INC"] = 4438] = "WINCOR_NIXDORF_INC";
    ManufacturerCode[ManufacturerCode["OHSUNG_ELECTRONICS"] = 4439] = "OHSUNG_ELECTRONICS";
    ManufacturerCode[ManufacturerCode["ZEN_WITHIN_INC"] = 4440] = "ZEN_WITHIN_INC";
    ManufacturerCode[ManufacturerCode["TECH4HOME_LDA"] = 4441] = "TECH4HOME_LDA";
    ManufacturerCode[ManufacturerCode["NANOLEAF"] = 4442] = "NANOLEAF";
    ManufacturerCode[ManufacturerCode["KEEN_HOME_INC"] = 4443] = "KEEN_HOME_INC";
    ManufacturerCode[ManufacturerCode["POLY_CONTROL_APS"] = 4444] = "POLY_CONTROL_APS";
    ManufacturerCode[ManufacturerCode["EASTFIELD_LIGHTING_CO_LTD_SHENZHEN"] = 4445] = "EASTFIELD_LIGHTING_CO_LTD_SHENZHEN";
    ManufacturerCode[ManufacturerCode["IP_DATATEL_INC"] = 4446] = "IP_DATATEL_INC";
    ManufacturerCode[ManufacturerCode["LUMI_UNITED_TECHOLOGY_LTD_SHENZHEN"] = 4447] = "LUMI_UNITED_TECHOLOGY_LTD_SHENZHEN";
    ManufacturerCode[ManufacturerCode["SENGLED_CO_LTD"] = 4448] = "SENGLED_CO_LTD";
    ManufacturerCode[ManufacturerCode["REMOTE_SOLUTION_CO_LTD"] = 4449] = "REMOTE_SOLUTION_CO_LTD";
    ManufacturerCode[ManufacturerCode["ABB_GENWAY_XIAMEN_ELECTRICAL_EQUIPMENT_CO_LTD"] = 4450] = "ABB_GENWAY_XIAMEN_ELECTRICAL_EQUIPMENT_CO_LTD";
    ManufacturerCode[ManufacturerCode["ZHEJIANG_REXENSE_TECH"] = 4451] = "ZHEJIANG_REXENSE_TECH";
    ManufacturerCode[ManufacturerCode["FOREE_TECHNOLOGY"] = 4452] = "FOREE_TECHNOLOGY";
    ManufacturerCode[ManufacturerCode["OPEN_ACCESS_TECHNOLOGY_INTL"] = 4453] = "OPEN_ACCESS_TECHNOLOGY_INTL";
    ManufacturerCode[ManufacturerCode["INNR_LIGHTING_BV"] = 4454] = "INNR_LIGHTING_BV";
    ManufacturerCode[ManufacturerCode["TECHWORLD_INDUSTRIES"] = 4455] = "TECHWORLD_INDUSTRIES";
    ManufacturerCode[ManufacturerCode["LEEDARSON_LIGHTING_CO_LTD"] = 4456] = "LEEDARSON_LIGHTING_CO_LTD";
    ManufacturerCode[ManufacturerCode["ARZEL_ZONING"] = 4457] = "ARZEL_ZONING";
    ManufacturerCode[ManufacturerCode["HOLLEY_TECHNOLOGY"] = 4458] = "HOLLEY_TECHNOLOGY";
    ManufacturerCode[ManufacturerCode["BELDON_TECHNOLOGIES"] = 4459] = "BELDON_TECHNOLOGIES";
    ManufacturerCode[ManufacturerCode["FLEXTRONICS"] = 4460] = "FLEXTRONICS";
    ManufacturerCode[ManufacturerCode["SHENZHEN_MEIAN"] = 4461] = "SHENZHEN_MEIAN";
    ManufacturerCode[ManufacturerCode["LOWES"] = 4462] = "LOWES";
    ManufacturerCode[ManufacturerCode["SIGMA_CONNECTIVITY"] = 4463] = "SIGMA_CONNECTIVITY";
    ManufacturerCode[ManufacturerCode["WULIAN"] = 4465] = "WULIAN";
    ManufacturerCode[ManufacturerCode["PLUGWISE_B_V"] = 4466] = "PLUGWISE_B_V";
    ManufacturerCode[ManufacturerCode["TITAN_PRODUCTS"] = 4467] = "TITAN_PRODUCTS";
    ManufacturerCode[ManufacturerCode["ECOSPECTRAL"] = 4468] = "ECOSPECTRAL";
    ManufacturerCode[ManufacturerCode["D_LINK"] = 4469] = "D_LINK";
    ManufacturerCode[ManufacturerCode["TECHNICOLOR_HOME_USA"] = 4470] = "TECHNICOLOR_HOME_USA";
    ManufacturerCode[ManufacturerCode["OPPLE_LIGHTING"] = 4471] = "OPPLE_LIGHTING";
    ManufacturerCode[ManufacturerCode["WISTRON_NEWEB_CORP"] = 4472] = "WISTRON_NEWEB_CORP";
    ManufacturerCode[ManufacturerCode["QMOTION_SHADES"] = 4473] = "QMOTION_SHADES";
    ManufacturerCode[ManufacturerCode["INSTA_GMBH"] = 4474] = "INSTA_GMBH";
    ManufacturerCode[ManufacturerCode["SHANGHAI_VANCOUNT"] = 4475] = "SHANGHAI_VANCOUNT";
    ManufacturerCode[ManufacturerCode["IKEA_OF_SWEDEN"] = 4476] = "IKEA_OF_SWEDEN";
    ManufacturerCode[ManufacturerCode["RT_RK"] = 4477] = "RT_RK";
    ManufacturerCode[ManufacturerCode["SHENZHEN_FEIBIT"] = 4478] = "SHENZHEN_FEIBIT";
    ManufacturerCode[ManufacturerCode["EUCONTROLS"] = 4479] = "EUCONTROLS";
    ManufacturerCode[ManufacturerCode["TELKONET"] = 4480] = "TELKONET";
    ManufacturerCode[ManufacturerCode["THERMAL_SOLUTION_RESOURCES"] = 4481] = "THERMAL_SOLUTION_RESOURCES";
    ManufacturerCode[ManufacturerCode["POMCUBE"] = 4482] = "POMCUBE";
    ManufacturerCode[ManufacturerCode["EI_ELECTRONICS"] = 4483] = "EI_ELECTRONICS";
    ManufacturerCode[ManufacturerCode["OPTOGA"] = 4484] = "OPTOGA";
    ManufacturerCode[ManufacturerCode["STELPRO"] = 4485] = "STELPRO";
    ManufacturerCode[ManufacturerCode["LYNXUS_TECHNOLOGIES_CORP"] = 4486] = "LYNXUS_TECHNOLOGIES_CORP";
    ManufacturerCode[ManufacturerCode["SEMICONDUCTOR_COMPONENTS"] = 4487] = "SEMICONDUCTOR_COMPONENTS";
    ManufacturerCode[ManufacturerCode["TP_LINK"] = 4488] = "TP_LINK";
    ManufacturerCode[ManufacturerCode["LEDVANCE_GMBH"] = 4489] = "LEDVANCE_GMBH";
    ManufacturerCode[ManufacturerCode["NORTEK"] = 4490] = "NORTEK";
    ManufacturerCode[ManufacturerCode["IREVO_ASSA_ABBLOY_KOREA"] = 4491] = "IREVO_ASSA_ABBLOY_KOREA";
    ManufacturerCode[ManufacturerCode["MIDEA"] = 4492] = "MIDEA";
    ManufacturerCode[ManufacturerCode["ZF_FRIEDRICHSHAFEN"] = 4493] = "ZF_FRIEDRICHSHAFEN";
    ManufacturerCode[ManufacturerCode["CHECKIT"] = 4494] = "CHECKIT";
    ManufacturerCode[ManufacturerCode["ACLARA"] = 4495] = "ACLARA";
    ManufacturerCode[ManufacturerCode["NOKIA"] = 4496] = "NOKIA";
    ManufacturerCode[ManufacturerCode["GOLDCARD_HIGH_TECH_CO_LTD"] = 4497] = "GOLDCARD_HIGH_TECH_CO_LTD";
    ManufacturerCode[ManufacturerCode["GEORGE_WILSON_INDUSTRIES_LTD"] = 4498] = "GEORGE_WILSON_INDUSTRIES_LTD";
    ManufacturerCode[ManufacturerCode["EASY_SAVER_CO_INC"] = 4499] = "EASY_SAVER_CO_INC";
    ManufacturerCode[ManufacturerCode["ZTE_CORPORATION"] = 4500] = "ZTE_CORPORATION";
    ManufacturerCode[ManufacturerCode["ARRIS"] = 4501] = "ARRIS";
    ManufacturerCode[ManufacturerCode["RELIANCE_BIG_TV"] = 4502] = "RELIANCE_BIG_TV";
    ManufacturerCode[ManufacturerCode["INSIGHT_ENERGY_VENTURES_POWERLEY"] = 4503] = "INSIGHT_ENERGY_VENTURES_POWERLEY";
    ManufacturerCode[ManufacturerCode["THOMAS_RESEARCH_PRODUCTS_HUBBELL_LIGHTING_INC"] = 4504] = "THOMAS_RESEARCH_PRODUCTS_HUBBELL_LIGHTING_INC";
    ManufacturerCode[ManufacturerCode["LI_SENG_TECHNOLOGY"] = 4505] = "LI_SENG_TECHNOLOGY";
    ManufacturerCode[ManufacturerCode["SYSTEM_LEVEL_SOLUTIONS_INC"] = 4506] = "SYSTEM_LEVEL_SOLUTIONS_INC";
    ManufacturerCode[ManufacturerCode["MATRIX_LABS"] = 4507] = "MATRIX_LABS";
    ManufacturerCode[ManufacturerCode["SINOPE_TECHNOLOGIES"] = 4508] = "SINOPE_TECHNOLOGIES";
    ManufacturerCode[ManufacturerCode["JIUZHOU_GREEBLE"] = 4509] = "JIUZHOU_GREEBLE";
    ManufacturerCode[ManufacturerCode["GUANGZHOU_LANVEE_TECH_CO_LTD"] = 4510] = "GUANGZHOU_LANVEE_TECH_CO_LTD";
    ManufacturerCode[ManufacturerCode["VENSTAR"] = 4511] = "VENSTAR";
    ManufacturerCode[ManufacturerCode["SLV"] = 4608] = "SLV";
    ManufacturerCode[ManufacturerCode["HALO_SMART_LABS"] = 4609] = "HALO_SMART_LABS";
    ManufacturerCode[ManufacturerCode["SCOUT_SECURITY_INC"] = 4610] = "SCOUT_SECURITY_INC";
    ManufacturerCode[ManufacturerCode["ALIBABA_CHINA_INC"] = 4611] = "ALIBABA_CHINA_INC";
    ManufacturerCode[ManufacturerCode["RESOLUTION_PRODUCTS_INC"] = 4612] = "RESOLUTION_PRODUCTS_INC";
    ManufacturerCode[ManufacturerCode["SMARTLOK_INC"] = 4613] = "SMARTLOK_INC";
    ManufacturerCode[ManufacturerCode["LUX_PRODUCTS_CORP"] = 4614] = "LUX_PRODUCTS_CORP";
    ManufacturerCode[ManufacturerCode["VIMAR_SPA"] = 4615] = "VIMAR_SPA";
    ManufacturerCode[ManufacturerCode["UNIVERSAL_LIGHTING_TECHNOLOGIES"] = 4616] = "UNIVERSAL_LIGHTING_TECHNOLOGIES";
    ManufacturerCode[ManufacturerCode["ROBERT_BOSCH_GMBH"] = 4617] = "ROBERT_BOSCH_GMBH";
    ManufacturerCode[ManufacturerCode["ACCENTURE"] = 4618] = "ACCENTURE";
    ManufacturerCode[ManufacturerCode["HEIMAN_TECHNOLOGY_CO_LTD"] = 4619] = "HEIMAN_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["SHENZHEN_HOMA_TECHNOLOGY_CO_LTD"] = 4620] = "SHENZHEN_HOMA_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["VISION_ELECTRONICS_TECHNOLOGY"] = 4621] = "VISION_ELECTRONICS_TECHNOLOGY";
    ManufacturerCode[ManufacturerCode["LENOVO"] = 4622] = "LENOVO";
    ManufacturerCode[ManufacturerCode["PRESCIENSE_RANDD"] = 4623] = "PRESCIENSE_RANDD";
    ManufacturerCode[ManufacturerCode["SHENZHEN_SEASTAR_INTELLIGENCE_CO_LTD"] = 4624] = "SHENZHEN_SEASTAR_INTELLIGENCE_CO_LTD";
    ManufacturerCode[ManufacturerCode["SENSATIVE_AB"] = 4625] = "SENSATIVE_AB";
    ManufacturerCode[ManufacturerCode["SOLAREDGE"] = 4626] = "SOLAREDGE";
    ManufacturerCode[ManufacturerCode["ZIPATO"] = 4627] = "ZIPATO";
    ManufacturerCode[ManufacturerCode["CHINA_FIRE_AND_SECURITY_SENSING_MANUFACTURING_IHORN"] = 4628] = "CHINA_FIRE_AND_SECURITY_SENSING_MANUFACTURING_IHORN";
    ManufacturerCode[ManufacturerCode["QUBY_BV"] = 4629] = "QUBY_BV";
    ManufacturerCode[ManufacturerCode["HANGZHOU_ROOMBANKER_TECHNOLOGY_CO_LTD"] = 4630] = "HANGZHOU_ROOMBANKER_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["AMAZON_LAB126"] = 4631] = "AMAZON_LAB126";
    ManufacturerCode[ManufacturerCode["PAULMANN_LICHT_GMBH"] = 4632] = "PAULMANN_LICHT_GMBH";
    ManufacturerCode[ManufacturerCode["SHENZHEN_ORVIBO_ELECTRONICS_CO_LTD"] = 4633] = "SHENZHEN_ORVIBO_ELECTRONICS_CO_LTD";
    ManufacturerCode[ManufacturerCode["TCI_TELECOMMUNICATIONS"] = 4634] = "TCI_TELECOMMUNICATIONS";
    ManufacturerCode[ManufacturerCode["MUELLER_LICHT_INTERNATIONAL_INC"] = 4635] = "MUELLER_LICHT_INTERNATIONAL_INC";
    ManufacturerCode[ManufacturerCode["AURORA_LIMITED"] = 4636] = "AURORA_LIMITED";
    ManufacturerCode[ManufacturerCode["SMARTDCC"] = 4637] = "SMARTDCC";
    ManufacturerCode[ManufacturerCode["SHANGHAI_UMEINFO_CO_LTD"] = 4638] = "SHANGHAI_UMEINFO_CO_LTD";
    ManufacturerCode[ManufacturerCode["CARBONTRACK"] = 4639] = "CARBONTRACK";
    ManufacturerCode[ManufacturerCode["SOMFY"] = 4640] = "SOMFY";
    ManufacturerCode[ManufacturerCode["VIESSMANN_ELEKTRONIK_GMBH"] = 4641] = "VIESSMANN_ELEKTRONIK_GMBH";
    ManufacturerCode[ManufacturerCode["HILDEBRAND_TECHNOLOGY_LTD"] = 4642] = "HILDEBRAND_TECHNOLOGY_LTD";
    ManufacturerCode[ManufacturerCode["ONKYO_TECHNOLOGY_CORPORATION"] = 4643] = "ONKYO_TECHNOLOGY_CORPORATION";
    ManufacturerCode[ManufacturerCode["SHENZHEN_SUNRICHER_TECHNOLOGY_LTD"] = 4644] = "SHENZHEN_SUNRICHER_TECHNOLOGY_LTD";
    ManufacturerCode[ManufacturerCode["XIU_XIU_TECHNOLOGY_CO_LTD"] = 4645] = "XIU_XIU_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["ZUMTOBEL_GROUP"] = 4646] = "ZUMTOBEL_GROUP";
    ManufacturerCode[ManufacturerCode["SHENZHEN_KAADAS_INTELLIGENT_TECHNOLOGY_CO_LTD"] = 4647] = "SHENZHEN_KAADAS_INTELLIGENT_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["SHANGHAI_XIAOYAN_TECHNOLOGY_CO_LTD"] = 4648] = "SHANGHAI_XIAOYAN_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["CYPRESS_SEMICONDUCTOR"] = 4649] = "CYPRESS_SEMICONDUCTOR";
    ManufacturerCode[ManufacturerCode["XAL_GMBH"] = 4650] = "XAL_GMBH";
    ManufacturerCode[ManufacturerCode["INERGY_SYSTEMS_LLC"] = 4651] = "INERGY_SYSTEMS_LLC";
    ManufacturerCode[ManufacturerCode["ALFRED_KARCHER_GMBH_AND_CO_KG"] = 4652] = "ALFRED_KARCHER_GMBH_AND_CO_KG";
    ManufacturerCode[ManufacturerCode["ADUROLIGHT_MANUFACTURING"] = 4653] = "ADUROLIGHT_MANUFACTURING";
    ManufacturerCode[ManufacturerCode["GROUPE_MULLER"] = 4654] = "GROUPE_MULLER";
    ManufacturerCode[ManufacturerCode["V_MARK_ENTERPRISES_INC"] = 4655] = "V_MARK_ENTERPRISES_INC";
    ManufacturerCode[ManufacturerCode["LEAD_ENERGY_AG"] = 4656] = "LEAD_ENERGY_AG";
    ManufacturerCode[ManufacturerCode["ULTIMATE_IOT_HENAN_TECHNOLOGY_LTD"] = 4657] = "ULTIMATE_IOT_HENAN_TECHNOLOGY_LTD";
    ManufacturerCode[ManufacturerCode["AXXESS_INDUSTRIES_INC"] = 4658] = "AXXESS_INDUSTRIES_INC";
    ManufacturerCode[ManufacturerCode["THIRD_REALITY_INC"] = 4659] = "THIRD_REALITY_INC";
    ManufacturerCode[ManufacturerCode["DSR_CORPORATION"] = 4660] = "DSR_CORPORATION";
    ManufacturerCode[ManufacturerCode["GUANGZHOU_VENSI_INTELLIGENT_TECHNOLOGY_CO_LTD"] = 4661] = "GUANGZHOU_VENSI_INTELLIGENT_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["SCHLAGE_LOCK_ALLEGION"] = 4662] = "SCHLAGE_LOCK_ALLEGION";
    ManufacturerCode[ManufacturerCode["NET2GRID"] = 4663] = "NET2GRID";
    ManufacturerCode[ManufacturerCode["AIRAM_ELECTRIC_OY_AB"] = 4664] = "AIRAM_ELECTRIC_OY_AB";
    ManufacturerCode[ManufacturerCode["IMMAX_WPB_CZ"] = 4665] = "IMMAX_WPB_CZ";
    ManufacturerCode[ManufacturerCode["ZIV_AUTOMATION"] = 4666] = "ZIV_AUTOMATION";
    ManufacturerCode[ManufacturerCode["HANGZHOU_IMAGICTECHNOLOGY_CO_LTD"] = 4667] = "HANGZHOU_IMAGICTECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["XIAMEN_LEELEN_TECHNOLOGY_CO_LTD"] = 4668] = "XIAMEN_LEELEN_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["OVERKIZ_SAS"] = 4669] = "OVERKIZ_SAS";
    ManufacturerCode[ManufacturerCode["FLONIDAN_A_S"] = 4670] = "FLONIDAN_A_S";
    ManufacturerCode[ManufacturerCode["HDL_AUTOMATION_CO_LTD"] = 4671] = "HDL_AUTOMATION_CO_LTD";
    ManufacturerCode[ManufacturerCode["ARDOMUS_NETWORKS_CORPORATION"] = 4672] = "ARDOMUS_NETWORKS_CORPORATION";
    ManufacturerCode[ManufacturerCode["SAMJIN_CO_LTD"] = 4673] = "SAMJIN_CO_LTD";
    ManufacturerCode[ManufacturerCode["FIREANGEL_SAFETY_TECHNOLOGY_PLC"] = 4674] = "FIREANGEL_SAFETY_TECHNOLOGY_PLC";
    ManufacturerCode[ManufacturerCode["INDRA_SISTEMAS_S_A"] = 4675] = "INDRA_SISTEMAS_S_A";
    ManufacturerCode[ManufacturerCode["SHENZHEN_JBT_SMART_LIGHTING_CO_LTD"] = 4676] = "SHENZHEN_JBT_SMART_LIGHTING_CO_LTD";
    ManufacturerCode[ManufacturerCode["GE_LIGHTING_AND_CURRENT"] = 4677] = "GE_LIGHTING_AND_CURRENT";
    ManufacturerCode[ManufacturerCode["DANFOSS_A_S"] = 4678] = "DANFOSS_A_S";
    ManufacturerCode[ManufacturerCode["NIVISS_PHP_SP_Z_O_O_SP_K"] = 4679] = "NIVISS_PHP_SP_Z_O_O_SP_K";
    ManufacturerCode[ManufacturerCode["SHENZHEN_FENGLIYUAN_ENERGY_CONSERVATING_TECHNOLOGY_CO_LTD"] = 4680] = "SHENZHEN_FENGLIYUAN_ENERGY_CONSERVATING_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["NEXELEC"] = 4681] = "NEXELEC";
    ManufacturerCode[ManufacturerCode["SICHUAN_BEHOME_PROMINENT_TECHNOLOGY_CO_LTD"] = 4682] = "SICHUAN_BEHOME_PROMINENT_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["FUJIAN_STAR_NET_COMMUNICATION_CO_LTD"] = 4683] = "FUJIAN_STAR_NET_COMMUNICATION_CO_LTD";
    ManufacturerCode[ManufacturerCode["TOSHIBA_VISUAL_SOLUTIONS_CORPORATION"] = 4684] = "TOSHIBA_VISUAL_SOLUTIONS_CORPORATION";
    ManufacturerCode[ManufacturerCode["LATCHABLE_INC"] = 4685] = "LATCHABLE_INC";
    ManufacturerCode[ManufacturerCode["LANDS_DEUTSCHLAND_GMBH"] = 4686] = "LANDS_DEUTSCHLAND_GMBH";
    ManufacturerCode[ManufacturerCode["GLEDOPTO_CO_LTD"] = 4687] = "GLEDOPTO_CO_LTD";
    ManufacturerCode[ManufacturerCode["THE_HOME_DEPOT"] = 4688] = "THE_HOME_DEPOT";
    ManufacturerCode[ManufacturerCode["NEONLITE_DISTRIBUTION_LIMITED"] = 4689] = "NEONLITE_DISTRIBUTION_LIMITED";
    ManufacturerCode[ManufacturerCode["ARLO_TECHNOLOGIES_INC"] = 4690] = "ARLO_TECHNOLOGIES_INC";
    ManufacturerCode[ManufacturerCode["XINGLUO_TECHNOLOGY_CO_LTD"] = 4691] = "XINGLUO_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["SIMON_ELECTRIC_CHINA_CO_LTD"] = 4692] = "SIMON_ELECTRIC_CHINA_CO_LTD";
    ManufacturerCode[ManufacturerCode["HANGZHOU_GREATSTAR_INDUSTRIAL_CO_LTD"] = 4693] = "HANGZHOU_GREATSTAR_INDUSTRIAL_CO_LTD";
    ManufacturerCode[ManufacturerCode["SEQUENTRIC_ENERGY_SYSTEMS_LLC"] = 4694] = "SEQUENTRIC_ENERGY_SYSTEMS_LLC";
    ManufacturerCode[ManufacturerCode["SOLUM_CO_LTD"] = 4695] = "SOLUM_CO_LTD";
    ManufacturerCode[ManufacturerCode["EAGLERISE_ELECTRIC_AND_ELECTRONIC_CHINA_CO_LTD"] = 4696] = "EAGLERISE_ELECTRIC_AND_ELECTRONIC_CHINA_CO_LTD";
    ManufacturerCode[ManufacturerCode["FANTEM_TECHNOLOGIES_SHENZHEN_CO_LTD"] = 4697] = "FANTEM_TECHNOLOGIES_SHENZHEN_CO_LTD";
    ManufacturerCode[ManufacturerCode["YUNDING_NETWORK_TECHNOLOGY_BEIJING_CO_LTD"] = 4698] = "YUNDING_NETWORK_TECHNOLOGY_BEIJING_CO_LTD";
    ManufacturerCode[ManufacturerCode["ATLANTIC_GROUP"] = 4699] = "ATLANTIC_GROUP";
    ManufacturerCode[ManufacturerCode["XIAMEN_INTRETECH_INC"] = 4700] = "XIAMEN_INTRETECH_INC";
    ManufacturerCode[ManufacturerCode["TUYA_GLOBAL_INC"] = 4701] = "TUYA_GLOBAL_INC";
    ManufacturerCode[ManufacturerCode["DNAKE_XIAMEN_INTELLIGENT_TECHNOLOGY_CO_LTD"] = 4702] = "DNAKE_XIAMEN_INTELLIGENT_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["NIKO_NV"] = 4703] = "NIKO_NV";
    ManufacturerCode[ManufacturerCode["EMPORIA_ENERGY"] = 4704] = "EMPORIA_ENERGY";
    ManufacturerCode[ManufacturerCode["SIKOM_AS"] = 4705] = "SIKOM_AS";
    ManufacturerCode[ManufacturerCode["AXIS_LABS_INC"] = 4706] = "AXIS_LABS_INC";
    ManufacturerCode[ManufacturerCode["CURRENT_PRODUCTS_CORPORATION"] = 4707] = "CURRENT_PRODUCTS_CORPORATION";
    ManufacturerCode[ManufacturerCode["METERSIT_SRL"] = 4708] = "METERSIT_SRL";
    ManufacturerCode[ManufacturerCode["HORNBACH_BAUMARKT_AG"] = 4709] = "HORNBACH_BAUMARKT_AG";
    ManufacturerCode[ManufacturerCode["DICEWORLD_S_R_L_A_SOCIO_UNICO"] = 4710] = "DICEWORLD_S_R_L_A_SOCIO_UNICO";
    ManufacturerCode[ManufacturerCode["ARC_TECHNOLOGY_CO_LTD"] = 4711] = "ARC_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["HANGZHOU_KONKE_INFORMATION_TECHNOLOGY_CO_LTD"] = 4712] = "HANGZHOU_KONKE_INFORMATION_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["SALTO_SYSTEMS_S_L"] = 4713] = "SALTO_SYSTEMS_S_L";
    ManufacturerCode[ManufacturerCode["SHENZHEN_SHYUGJ_TECHNOLOGY_CO_LTD"] = 4714] = "SHENZHEN_SHYUGJ_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["BRAYDEN_AUTOMATION_CORPORATION"] = 4715] = "BRAYDEN_AUTOMATION_CORPORATION";
    ManufacturerCode[ManufacturerCode["ENVIRONEXUS_PTY_LTD"] = 4716] = "ENVIRONEXUS_PTY_LTD";
    ManufacturerCode[ManufacturerCode["ELTRA_NV_SA"] = 4717] = "ELTRA_NV_SA";
    ManufacturerCode[ManufacturerCode["XIAOMI_COMMUNICATIONS_CO_LTD"] = 4718] = "XIAOMI_COMMUNICATIONS_CO_LTD";
    ManufacturerCode[ManufacturerCode["SHANGHAI_SHUNCOM_ELECTRONIC_TECHNOLOGY_CO_LTD"] = 4719] = "SHANGHAI_SHUNCOM_ELECTRONIC_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["VOLTALIS_S_A"] = 4720] = "VOLTALIS_S_A";
    ManufacturerCode[ManufacturerCode["FEELUX_CO_LTD"] = 4721] = "FEELUX_CO_LTD";
    ManufacturerCode[ManufacturerCode["SMARTPLUS_INC"] = 4722] = "SMARTPLUS_INC";
    ManufacturerCode[ManufacturerCode["HALEMEIER_GMBH"] = 4723] = "HALEMEIER_GMBH";
    ManufacturerCode[ManufacturerCode["TRUST_INTERNATIONAL_BV"] = 4724] = "TRUST_INTERNATIONAL_BV";
    ManufacturerCode[ManufacturerCode["DUKE_ENERGY_BUSINESS_SERVICES_LLC"] = 4725] = "DUKE_ENERGY_BUSINESS_SERVICES_LLC";
    ManufacturerCode[ManufacturerCode["CALIX_INC"] = 4726] = "CALIX_INC";
    ManufacturerCode[ManufacturerCode["ADEO"] = 4727] = "ADEO";
    ManufacturerCode[ManufacturerCode["CONNECTED_RESPONSE_LIMITED"] = 4728] = "CONNECTED_RESPONSE_LIMITED";
    ManufacturerCode[ManufacturerCode["STROYENERGOKOM_LTD"] = 4729] = "STROYENERGOKOM_LTD";
    ManufacturerCode[ManufacturerCode["LUMITECH_LIGHTING_SOLUTION_GMBH"] = 4730] = "LUMITECH_LIGHTING_SOLUTION_GMBH";
    ManufacturerCode[ManufacturerCode["VERDANT_ENVIRONMENTAL_TECHNOLOGIES"] = 4731] = "VERDANT_ENVIRONMENTAL_TECHNOLOGIES";
    ManufacturerCode[ManufacturerCode["ALFRED_INTERNATIONAL_INC"] = 4732] = "ALFRED_INTERNATIONAL_INC";
    ManufacturerCode[ManufacturerCode["SANSI_LED_LIGHTING_CO_LTD"] = 4733] = "SANSI_LED_LIGHTING_CO_LTD";
    ManufacturerCode[ManufacturerCode["MINDTREE_LIMITED"] = 4734] = "MINDTREE_LIMITED";
    ManufacturerCode[ManufacturerCode["NORDIC_SEMICONDUCTOR_ASA"] = 4735] = "NORDIC_SEMICONDUCTOR_ASA";
    ManufacturerCode[ManufacturerCode["SITERWELL_ELECTRONICS_CO_LIMITED"] = 4736] = "SITERWELL_ELECTRONICS_CO_LIMITED";
    ManufacturerCode[ManufacturerCode["BRILONER_LEUCHTEN_GMBH_AND_CO_KG"] = 4737] = "BRILONER_LEUCHTEN_GMBH_AND_CO_KG";
    ManufacturerCode[ManufacturerCode["SHENZHEN_SEI_TECHNOLOGY_CO_LTD"] = 4738] = "SHENZHEN_SEI_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["COPPER_LABS_INC"] = 4739] = "COPPER_LABS_INC";
    ManufacturerCode[ManufacturerCode["DELTA_DORE"] = 4740] = "DELTA_DORE";
    ManufacturerCode[ManufacturerCode["HAGER_GROUP"] = 4741] = "HAGER_GROUP";
    ManufacturerCode[ManufacturerCode["SHENZHEN_COOLKIT_TECHNOLOGY_CO_LTD"] = 4742] = "SHENZHEN_COOLKIT_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["HANGZHOU_SKY_LIGHTING_CO_LTD"] = 4743] = "HANGZHOU_SKY_LIGHTING_CO_LTD";
    ManufacturerCode[ManufacturerCode["E_ON_SE"] = 4744] = "E_ON_SE";
    ManufacturerCode[ManufacturerCode["LIDL_STIFTUNG_AND_CO_KG"] = 4745] = "LIDL_STIFTUNG_AND_CO_KG";
    ManufacturerCode[ManufacturerCode["SICHUAN_CHANGHONG_NETWORK_TECHNOLOGIES_CO_LTD"] = 4746] = "SICHUAN_CHANGHONG_NETWORK_TECHNOLOGIES_CO_LTD";
    ManufacturerCode[ManufacturerCode["NODON"] = 4747] = "NODON";
    ManufacturerCode[ManufacturerCode["JIANGXI_INNOTECH_TECHNOLOGY_CO_LTD"] = 4748] = "JIANGXI_INNOTECH_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["MERCATOR_PTY_LTD"] = 4749] = "MERCATOR_PTY_LTD";
    ManufacturerCode[ManufacturerCode["BEIJING_RUYING_TECH_LIMITED"] = 4750] = "BEIJING_RUYING_TECH_LIMITED";
    ManufacturerCode[ManufacturerCode["EGLO_LEUCHTEN_GMBH"] = 4751] = "EGLO_LEUCHTEN_GMBH";
    ManufacturerCode[ManufacturerCode["PIETRO_FIORENTINI_S_P_A"] = 4752] = "PIETRO_FIORENTINI_S_P_A";
    ManufacturerCode[ManufacturerCode["ZEHNDER_GROUP_VAUX_ANDIGNY"] = 4753] = "ZEHNDER_GROUP_VAUX_ANDIGNY";
    ManufacturerCode[ManufacturerCode["BRK_BRANDS_INC"] = 4754] = "BRK_BRANDS_INC";
    ManufacturerCode[ManufacturerCode["ASKEY_COMPUTER_CORP"] = 4755] = "ASKEY_COMPUTER_CORP";
    ManufacturerCode[ManufacturerCode["PASSIVEBOLT_INC"] = 4756] = "PASSIVEBOLT_INC";
    ManufacturerCode[ManufacturerCode["AVM_AUDIOVISUELLES_MARKETING_UND_COMPUTERSYSTEME_BERLIN"] = 4757] = "AVM_AUDIOVISUELLES_MARKETING_UND_COMPUTERSYSTEME_BERLIN";
    ManufacturerCode[ManufacturerCode["NINGBO_SUNTECH_LIGHTING_TECHNOLOGY_CO_LTD"] = 4758] = "NINGBO_SUNTECH_LIGHTING_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["SOCIETE_EN_COMMANDITE_STELLO"] = 4759] = "SOCIETE_EN_COMMANDITE_STELLO";
    ManufacturerCode[ManufacturerCode["VIVINT_SMART_HOME"] = 4760] = "VIVINT_SMART_HOME";
    ManufacturerCode[ManufacturerCode["NAMRON_AS"] = 4761] = "NAMRON_AS";
    ManufacturerCode[ManufacturerCode["RADEMACHER_GERAETE_ELEKTRONIK_GMBH"] = 4762] = "RADEMACHER_GERAETE_ELEKTRONIK_GMBH";
    ManufacturerCode[ManufacturerCode["OMO_SYSTEMS_LTD"] = 4763] = "OMO_SYSTEMS_LTD";
    ManufacturerCode[ManufacturerCode["SIGLIS_AG"] = 4764] = "SIGLIS_AG";
    ManufacturerCode[ManufacturerCode["IMHOTEP_CREATION"] = 4765] = "IMHOTEP_CREATION";
    ManufacturerCode[ManufacturerCode["ICASA"] = 4766] = "ICASA";
    ManufacturerCode[ManufacturerCode["LEVEL_HOME_INC"] = 4767] = "LEVEL_HOME_INC";
    ManufacturerCode[ManufacturerCode["TIS_CONTROL_LIMITED"] = 4864] = "TIS_CONTROL_LIMITED";
    ManufacturerCode[ManufacturerCode["RADISYS_INDIA_PVT_LTD"] = 4865] = "RADISYS_INDIA_PVT_LTD";
    ManufacturerCode[ManufacturerCode["VEEA_INC"] = 4866] = "VEEA_INC";
    ManufacturerCode[ManufacturerCode["FELL_TECHNOLOGY_AS"] = 4867] = "FELL_TECHNOLOGY_AS";
    ManufacturerCode[ManufacturerCode["SOWILO_DESIGN_SERVICES_LTD"] = 4868] = "SOWILO_DESIGN_SERVICES_LTD";
    ManufacturerCode[ManufacturerCode["LEXI_DEVICES_INC"] = 4869] = "LEXI_DEVICES_INC";
    ManufacturerCode[ManufacturerCode["LIFI_LABS_INC_DBA_LIFX"] = 4870] = "LIFI_LABS_INC_DBA_LIFX";
    ManufacturerCode[ManufacturerCode["GRUNDFOS_HOLDING_A_S"] = 4871] = "GRUNDFOS_HOLDING_A_S";
    ManufacturerCode[ManufacturerCode["SOURCING_AND_CREATION"] = 4872] = "SOURCING_AND_CREATION";
    ManufacturerCode[ManufacturerCode["KRAKEN_TECHNOLOGIES_LTD"] = 4873] = "KRAKEN_TECHNOLOGIES_LTD";
    ManufacturerCode[ManufacturerCode["EVE_SYSTEMS"] = 4874] = "EVE_SYSTEMS";
    ManufacturerCode[ManufacturerCode["LITE_ON_TECHNOLOGY_CORPORATION"] = 4875] = "LITE_ON_TECHNOLOGY_CORPORATION";
    ManufacturerCode[ManufacturerCode["FOCALCREST_LIMITED"] = 4876] = "FOCALCREST_LIMITED";
    ManufacturerCode[ManufacturerCode["BOUFFALO_LAB_NANJING_CO_LTD"] = 4877] = "BOUFFALO_LAB_NANJING_CO_LTD";
    ManufacturerCode[ManufacturerCode["WYZE_LABS_INC"] = 4878] = "WYZE_LABS_INC";
    ManufacturerCode[ManufacturerCode["Z_WAVE_EUROPE_GMBH"] = 4879] = "Z_WAVE_EUROPE_GMBH";
    ManufacturerCode[ManufacturerCode["AEOTEC_LIMITED"] = 4880] = "AEOTEC_LIMITED";
    ManufacturerCode[ManufacturerCode["NGSTB_COMPANY_LIMITED"] = 4881] = "NGSTB_COMPANY_LIMITED";
    ManufacturerCode[ManufacturerCode["QINGDAO_YEELINK_INFORMATION_TECHNOLOGY_CO_LTD"] = 4882] = "QINGDAO_YEELINK_INFORMATION_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["E_SMART_HOME_AUTOMATION_SYSTEMS_LIMITED"] = 4883] = "E_SMART_HOME_AUTOMATION_SYSTEMS_LIMITED";
    ManufacturerCode[ManufacturerCode["FIBAR_GROUP_S_A"] = 4884] = "FIBAR_GROUP_S_A";
    ManufacturerCode[ManufacturerCode["PROLITECH_GMBH"] = 4885] = "PROLITECH_GMBH";
    ManufacturerCode[ManufacturerCode["PANKORE_INTEGRATED_CIRCUIT_TECHNOLOGY_CO_LTD"] = 4886] = "PANKORE_INTEGRATED_CIRCUIT_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["LOGITECH"] = 4887] = "LOGITECH";
    ManufacturerCode[ManufacturerCode["PIARO_INC"] = 4888] = "PIARO_INC";
    ManufacturerCode[ManufacturerCode["MITSUBISHI_ELECTRIC_US_INC"] = 4889] = "MITSUBISHI_ELECTRIC_US_INC";
    ManufacturerCode[ManufacturerCode["RESIDEO_TECHNOLOGIES_INC"] = 4890] = "RESIDEO_TECHNOLOGIES_INC";
    ManufacturerCode[ManufacturerCode["ESPRESSIF_SYSTEMS_SHANGHAI_CO_LTD"] = 4891] = "ESPRESSIF_SYSTEMS_SHANGHAI_CO_LTD";
    ManufacturerCode[ManufacturerCode["HELLA_SONNEN_UND_WETTERSCHUTZTECHNIK_GMBH"] = 4892] = "HELLA_SONNEN_UND_WETTERSCHUTZTECHNIK_GMBH";
    ManufacturerCode[ManufacturerCode["GEBERIT_INTERNATIONAL_AG"] = 4893] = "GEBERIT_INTERNATIONAL_AG";
    ManufacturerCode[ManufacturerCode["CAME_S_P_A"] = 4894] = "CAME_S_P_A";
    ManufacturerCode[ManufacturerCode["GUANGZHOU_ELITE_EDUCATION_AND_TECHNOLOGY_CO_LTD"] = 4895] = "GUANGZHOU_ELITE_EDUCATION_AND_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["PHYPLUS_MICROELECTRONICS_LIMITED"] = 4896] = "PHYPLUS_MICROELECTRONICS_LIMITED";
    ManufacturerCode[ManufacturerCode["SHENZHEN_SONOFF_TECHNOLOGIES_CO_LTD"] = 4897] = "SHENZHEN_SONOFF_TECHNOLOGIES_CO_LTD";
    ManufacturerCode[ManufacturerCode["SAFE4_SECURITY_GROUP"] = 4898] = "SAFE4_SECURITY_GROUP";
    ManufacturerCode[ManufacturerCode["SHANGHAI_MXCHIP_INFORMATION_TECHNOLOGY_CO_LTD"] = 4899] = "SHANGHAI_MXCHIP_INFORMATION_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["HDC_I_CONTROLS"] = 4900] = "HDC_I_CONTROLS";
    ManufacturerCode[ManufacturerCode["ZUMA_ARRAY_LIMITED"] = 4901] = "ZUMA_ARRAY_LIMITED";
    ManufacturerCode[ManufacturerCode["DECELECT"] = 4902] = "DECELECT";
    ManufacturerCode[ManufacturerCode["MILL_INTERNATIONAL_AS"] = 4903] = "MILL_INTERNATIONAL_AS";
    ManufacturerCode[ManufacturerCode["HOMEWIZARD_BV"] = 4904] = "HOMEWIZARD_BV";
    ManufacturerCode[ManufacturerCode["SHENZHEN_TOPBAND_CO_LTD"] = 4905] = "SHENZHEN_TOPBAND_CO_LTD";
    ManufacturerCode[ManufacturerCode["PRESSAC_COMMUNICATIONS_LTD"] = 4906] = "PRESSAC_COMMUNICATIONS_LTD";
    ManufacturerCode[ManufacturerCode["ORIGIN_WIRELESS_INC"] = 4907] = "ORIGIN_WIRELESS_INC";
    ManufacturerCode[ManufacturerCode["CONNECTE_AS"] = 4908] = "CONNECTE_AS";
    ManufacturerCode[ManufacturerCode["YOKIS"] = 4909] = "YOKIS";
    ManufacturerCode[ManufacturerCode["XIAMEN_YANKON_ENERGETIC_LIGHTING_CO_LTD"] = 4910] = "XIAMEN_YANKON_ENERGETIC_LIGHTING_CO_LTD";
    ManufacturerCode[ManufacturerCode["YANDEX_LLC"] = 4911] = "YANDEX_LLC";
    ManufacturerCode[ManufacturerCode["CRITICAL_SOFTWARE_S_A"] = 4912] = "CRITICAL_SOFTWARE_S_A";
    ManufacturerCode[ManufacturerCode["NORTEK_CONTROL"] = 4913] = "NORTEK_CONTROL";
    ManufacturerCode[ManufacturerCode["BRIGHTAI"] = 4914] = "BRIGHTAI";
    ManufacturerCode[ManufacturerCode["BECKER_ANTRIEBE_GMBH"] = 4915] = "BECKER_ANTRIEBE_GMBH";
    ManufacturerCode[ManufacturerCode["SHENZHEN_TCL_NEW_TECHNOLOGY_COMPANY_LIMITED"] = 4916] = "SHENZHEN_TCL_NEW_TECHNOLOGY_COMPANY_LIMITED";
    ManufacturerCode[ManufacturerCode["DEXATEK_TECHNOLOGY_LTD"] = 4917] = "DEXATEK_TECHNOLOGY_LTD";
    ManufacturerCode[ManufacturerCode["ELELABS_INTERNATIONAL_LIMITED"] = 4918] = "ELELABS_INTERNATIONAL_LIMITED";
    ManufacturerCode[ManufacturerCode["DATEK_WIRELESS_AS"] = 4919] = "DATEK_WIRELESS_AS";
    ManufacturerCode[ManufacturerCode["ALDES"] = 4920] = "ALDES";
    ManufacturerCode[ManufacturerCode["SAVANT_COMPANY"] = 4921] = "SAVANT_COMPANY";
    ManufacturerCode[ManufacturerCode["ARISTON_THERMO_GROUP"] = 4922] = "ARISTON_THERMO_GROUP";
    ManufacturerCode[ManufacturerCode["WAREMA_RENKHOFF_SE"] = 4923] = "WAREMA_RENKHOFF_SE";
    ManufacturerCode[ManufacturerCode["VTECH_HOLDINGS_LIMITED"] = 4924] = "VTECH_HOLDINGS_LIMITED";
    ManufacturerCode[ManufacturerCode["FUTUREHOME_AS"] = 4925] = "FUTUREHOME_AS";
    ManufacturerCode[ManufacturerCode["COGNITIVE_SYSTEMS_CORP"] = 4926] = "COGNITIVE_SYSTEMS_CORP";
    ManufacturerCode[ManufacturerCode["ASR_MICROELECTRONICS_SHENZHEN_CO_LTD"] = 4927] = "ASR_MICROELECTRONICS_SHENZHEN_CO_LTD";
    ManufacturerCode[ManufacturerCode["AIRIOS"] = 4928] = "AIRIOS";
    ManufacturerCode[ManufacturerCode["GUANGDONG_OPPO_MOBILE_TELECOMMUNICATIONS_CORP_LTD"] = 4929] = "GUANGDONG_OPPO_MOBILE_TELECOMMUNICATIONS_CORP_LTD";
    ManufacturerCode[ManufacturerCode["BEKEN_CORPORATION"] = 4930] = "BEKEN_CORPORATION";
    ManufacturerCode[ManufacturerCode["CORSAIR"] = 4931] = "CORSAIR";
    ManufacturerCode[ManufacturerCode["ELTAKO_GMBH"] = 4932] = "ELTAKO_GMBH";
    ManufacturerCode[ManufacturerCode["CHENGDU_MEROSS_TECHNOLOGY_CO_LTD"] = 4933] = "CHENGDU_MEROSS_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["RAFAEL_MICROELECTRONICS_INC"] = 4934] = "RAFAEL_MICROELECTRONICS_INC";
    ManufacturerCode[ManufacturerCode["AUG_WINKHUAS_GMBH_AND_CO_KG"] = 4935] = "AUG_WINKHUAS_GMBH_AND_CO_KG";
    ManufacturerCode[ManufacturerCode["QINGDAO_HAIER_TECHNOLOGY_CO_LTD"] = 4936] = "QINGDAO_HAIER_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["APPLE_INC"] = 4937] = "APPLE_INC";
    ManufacturerCode[ManufacturerCode["ROLLEASE_ACMEDA"] = 4938] = "ROLLEASE_ACMEDA";
    ManufacturerCode[ManufacturerCode["NABU_CASA_INC"] = 4939] = "NABU_CASA_INC";
    ManufacturerCode[ManufacturerCode["SIMON_HOLDING"] = 4940] = "SIMON_HOLDING";
    ManufacturerCode[ManufacturerCode["KD_NAVIEN"] = 4941] = "KD_NAVIEN";
    ManufacturerCode[ManufacturerCode["TADO_GMBH"] = 4942] = "TADO_GMBH";
    ManufacturerCode[ManufacturerCode["MEDIOLA_CONNECTED_LIVING_AG"] = 4943] = "MEDIOLA_CONNECTED_LIVING_AG";
    ManufacturerCode[ManufacturerCode["POLYNHOME"] = 4944] = "POLYNHOME";
    ManufacturerCode[ManufacturerCode["HOORII_TECHNOLOGY_CO_LTD"] = 4945] = "HOORII_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["KIMIN_ELECTRONICS_CO_LTD"] = 4947] = "KIMIN_ELECTRONICS_CO_LTD";
    ManufacturerCode[ManufacturerCode["ZYAX_AB"] = 4948] = "ZYAX_AB";
    ManufacturerCode[ManufacturerCode["BARACODA_SA"] = 4949] = "BARACODA_SA";
    ManufacturerCode[ManufacturerCode["LENNOX_INTERNATIONAL_INC"] = 4950] = "LENNOX_INTERNATIONAL_INC";
    ManufacturerCode[ManufacturerCode["TELEDATICS_INCORPORATED"] = 4951] = "TELEDATICS_INCORPORATED";
    ManufacturerCode[ManufacturerCode["TOP_VICTORY_INVESTMENTS_LIMITED"] = 4952] = "TOP_VICTORY_INVESTMENTS_LIMITED";
    ManufacturerCode[ManufacturerCode["GOQUAL_INC"] = 4953] = "GOQUAL_INC";
    ManufacturerCode[ManufacturerCode["SIEGENIA_AUBI_KG"] = 4954] = "SIEGENIA_AUBI_KG";
    ManufacturerCode[ManufacturerCode["VIRTUAL_CONNECTED_CONTROLLING_SYSTEM_SINGAPORE_PTE_LTD"] = 4955] = "VIRTUAL_CONNECTED_CONTROLLING_SYSTEM_SINGAPORE_PTE_LTD";
    ManufacturerCode[ManufacturerCode["GIGASET_COMMUNICATIONS_GMBH"] = 4956] = "GIGASET_COMMUNICATIONS_GMBH";
    ManufacturerCode[ManufacturerCode["NUKI_HOME_SOLUTIONS_GMBH"] = 4957] = "NUKI_HOME_SOLUTIONS_GMBH";
    ManufacturerCode[ManufacturerCode["DEVICEBOOK_INC"] = 4958] = "DEVICEBOOK_INC";
    ManufacturerCode[ManufacturerCode["CONSUMER_2_0_INC_RENTLY"] = 4959] = "CONSUMER_2_0_INC_RENTLY";
    ManufacturerCode[ManufacturerCode["EDISON_LABS_INC_DBA_ORRO"] = 4960] = "EDISON_LABS_INC_DBA_ORRO";
    ManufacturerCode[ManufacturerCode["INOVELLI"] = 4961] = "INOVELLI";
    ManufacturerCode[ManufacturerCode["DEVERITEC_GMBH"] = 4962] = "DEVERITEC_GMBH";
    ManufacturerCode[ManufacturerCode["CHARTER_COMMUNICATIONS"] = 4963] = "CHARTER_COMMUNICATIONS";
    ManufacturerCode[ManufacturerCode["MONOLITHIC_POWER_SYSTEMS_INC"] = 4964] = "MONOLITHIC_POWER_SYSTEMS_INC";
    ManufacturerCode[ManufacturerCode["NINGBO_DOOYA_MECHANIC_AND_ELECTRONIC_TECHNOLOGY_CO_LTD"] = 4965] = "NINGBO_DOOYA_MECHANIC_AND_ELECTRONIC_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["SHENZHEN_SDMC_TECHNOLOGY_CO_LTD"] = 4966] = "SHENZHEN_SDMC_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["HP_INC"] = 4967] = "HP_INC";
    ManufacturerCode[ManufacturerCode["MUI_LAB_INC"] = 4968] = "MUI_LAB_INC";
    ManufacturerCode[ManufacturerCode["BHTRONICS_S_R_L"] = 4969] = "BHTRONICS_S_R_L";
    ManufacturerCode[ManufacturerCode["AKUVOX_XIAMEN_NETWORKS_CO_LTD"] = 4970] = "AKUVOX_XIAMEN_NETWORKS_CO_LTD";
    ManufacturerCode[ManufacturerCode["SHELLY"] = 5264] = "SHELLY";
    ManufacturerCode[ManufacturerCode["SBERDEVICES_LTD"] = 5423] = "SBERDEVICES_LTD";
    ManufacturerCode[ManufacturerCode["GEWISS_S_P_A"] = 6548] = "GEWISS_S_P_A";
    ManufacturerCode[ManufacturerCode["CLIMAX_TECHNOLOGY_CO_LTD"] = 10132] = "CLIMAX_TECHNOLOGY_CO_LTD";
    ManufacturerCode[ManufacturerCode["GOOGLE_LLC"] = 24582] = "GOOGLE_LLC";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE"] = 50592] = "CONNECTIVITY_STANDARDS_ALLIANCE";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_2"] = 50593] = "CONNECTIVITY_STANDARDS_ALLIANCE_2";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_3"] = 50594] = "CONNECTIVITY_STANDARDS_ALLIANCE_3";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_4"] = 50595] = "CONNECTIVITY_STANDARDS_ALLIANCE_4";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_5"] = 50596] = "CONNECTIVITY_STANDARDS_ALLIANCE_5";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_6"] = 50597] = "CONNECTIVITY_STANDARDS_ALLIANCE_6";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_7"] = 50598] = "CONNECTIVITY_STANDARDS_ALLIANCE_7";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_8"] = 50599] = "CONNECTIVITY_STANDARDS_ALLIANCE_8";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_9"] = 50600] = "CONNECTIVITY_STANDARDS_ALLIANCE_9";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_10"] = 50601] = "CONNECTIVITY_STANDARDS_ALLIANCE_10";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_11"] = 50602] = "CONNECTIVITY_STANDARDS_ALLIANCE_11";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_12"] = 50603] = "CONNECTIVITY_STANDARDS_ALLIANCE_12";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_13"] = 50604] = "CONNECTIVITY_STANDARDS_ALLIANCE_13";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_14"] = 50605] = "CONNECTIVITY_STANDARDS_ALLIANCE_14";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_15"] = 50606] = "CONNECTIVITY_STANDARDS_ALLIANCE_15";
    ManufacturerCode[ManufacturerCode["CONNECTIVITY_STANDARDS_ALLIANCE_16"] = 50607] = "CONNECTIVITY_STANDARDS_ALLIANCE_16";
    ManufacturerCode[ManufacturerCode["TEST_VENDOR_1"] = 65521] = "TEST_VENDOR_1";
    ManufacturerCode[ManufacturerCode["TEST_VENDOR_2"] = 65522] = "TEST_VENDOR_2";
    ManufacturerCode[ManufacturerCode["TEST_VENDOR_3"] = 65523] = "TEST_VENDOR_3";
    ManufacturerCode[ManufacturerCode["TEST_VENDOR_4"] = 65524] = "TEST_VENDOR_4";
    ManufacturerCode[ManufacturerCode["RESERVED"] = 65525] = "RESERVED";
    ManufacturerCode[ManufacturerCode["RESERVED_1"] = 65526] = "RESERVED_1";
    ManufacturerCode[ManufacturerCode["RESERVED_2"] = 65527] = "RESERVED_2";
    ManufacturerCode[ManufacturerCode["RESERVED_3"] = 65528] = "RESERVED_3";
    ManufacturerCode[ManufacturerCode["RESERVED_4"] = 65529] = "RESERVED_4";
    ManufacturerCode[ManufacturerCode["RESERVED_5"] = 65530] = "RESERVED_5";
    ManufacturerCode[ManufacturerCode["RESERVED_6"] = 65531] = "RESERVED_6";
    ManufacturerCode[ManufacturerCode["RESERVED_7"] = 65532] = "RESERVED_7";
    ManufacturerCode[ManufacturerCode["RESERVED_8"] = 65533] = "RESERVED_8";
    ManufacturerCode[ManufacturerCode["RESERVED_9"] = 65534] = "RESERVED_9";
    ManufacturerCode[ManufacturerCode["RESERVED_10"] = 65535] = "RESERVED_10";
    // Non-CSA-assigned codes
    ManufacturerCode[ManufacturerCode["CUSTOM_PERENIO"] = 123] = "CUSTOM_PERENIO";
    ManufacturerCode[ManufacturerCode["CUSTOM_LIVOLO"] = 6866] = "CUSTOM_LIVOLO";
    ManufacturerCode[ManufacturerCode["CUSTOM_SPRUT_DEVICE"] = 26214] = "CUSTOM_SPRUT_DEVICE";
    ManufacturerCode[ManufacturerCode["CUSTOM_LYTKO"] = 30583] = "CUSTOM_LYTKO";
})(ManufacturerCode || (exports.ManufacturerCode = ManufacturerCode = {}));
